{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 1 - _Foundations of Information Retrieval 2022_\n",
    "\n",
    "This assignment is divided in 4 parts, which have to be delivered all together no later than 03/10/2022 (strict - no extensions will be granted!), via Canvas. Delivery of the assignment solutions is mandatory (_see grading conditions on Canvas and in slides of Lecture01_).\n",
    "\n",
    "We will use [ElasticSearch](https://www.elastic.co/) as search engine. It provides state-of-the-art tools to implement your own engine, index your documents, and let you focus on methodological aspects of search models and optimization. \n",
    "\n",
    "The assignment is about text-based Information Retrieval and it is structured in three parts:\n",
    "1. IR performance evaluation (implementation of performance metrics)\n",
    "2. Setting up a search engine, pre-processing and indexing using ElasticSearch (Indexing, Analyzers)\n",
    "3. Implementation and optimization of models of search (Similarity)\n",
    "\n",
    "\n",
    "This assignment contains exercises, marked with the section title __Exercise 01.(x)__, which are evaluated, and other sections that contain support code which you should study and use as it is. Write your answers between the comments `BEGIN ANSWER` and `END ANSWER`. \n",
    "\n",
    "_Note:_ the comment `#THIS IS GRADED!` in a section indicates that it will be graded.\n",
    "\n",
    "\n",
    "### Initial preparation (self-study)\n",
    "For the first part, it is good to acquire (or refresh) basic knowledge of Python. Please use the [Python tutorials](https://docs.python.org/3/tutorial/) if needed.\n",
    "\n",
    "For the second and third part of the assignment, please study yourself the [Getting Started guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/getting-started.html) of ElasticSearch and get acquainted with the framework.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "***\n",
    "\n",
    "# PART 01 - Performance evaluation\n",
    "\n",
    "\n",
    "### Background information and reading\n",
    "Study the slides of Lecture 01 (available on Canvas) and the reference book chapter (Christopher D. Manning, Prabhakar Raghavan and Hinrich Schütze, Introduction to Information Retrieval, [Chapter 8, Evaluation in information retrieval](http://nlp.stanford.edu/IR-book/pdf/08eval.pdf), Cambridge University Press. 2008)\n",
    "\n",
    "### Basic concepts\n",
    "Suppose the set of relevant documents (the document identifiers - _doc-IDs_) is called `relevant`, then we  define it as follows (in Python):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "relevant = set([2, 3, 5, 8, 13, 17, 21, 34, 38])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A perfect run would retrieve exactly these 9 documents in any order. Now, suppose the list of retrieved documents (the document identifiers - _doc-IDs_) is called `retrieved`, and contains the following _doc-IDs_:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "retrieved = [14, 4, 2, 18, 16, 8, 46, 32, 17, 34, 33, 22, 47, 39, 11]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of the simplest evaluation measures is the _Success at rank 1_, i.e. `Is the first document retrieved a relevant document?`\n",
    "\n",
    "_Success at rank 1_ returns 1 if the first document is relevant, and 0 otherwise. A possible implementation is: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def success_at_1 (relevant, retrieved):\n",
    "    if len(retrieved) > 0 and retrieved[0] in relevant:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "success_at_1(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first retrieved documentid is 14 which is not in the set of relevant documents, so the `success_at_1` is 0.\n",
    "\n",
    "_________________\n",
    "\n",
    "> Note how easy it is to check if an item occurs in a Python set or list by using the keyword: `in`. Similarly, you can loop over all items in a set of list with: \n",
    "`for doc in retrieved:`, \n",
    "where doc will refer to each item in the set or list. \n",
    "\n",
    "Be sure to use the internet to sharpen your knowledge about Python constructs, for instance on [Python list slicing](https://duckduckgo.com/?q=python+list+slicing). Also note that the code above checks if at least one document is retrieved to avoid an index out of bounds exception (i.e. we avoid to access an empty vector).\n",
    "\n",
    "> ___Suggestion:___ _to be sure of the correctness of the implementation of the performance metrics, you can compute their values manually and compare them with those computed by your functions. This is important, as you will use these metrics for later exercises and to compare the results of differentmodels._"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation exercise: _Success at k_\n",
    "The measure _Success at k_ returns 1 if a relevant document is among the first _k_ documents retrieved and zero otherwise.\n",
    "\n",
    "> Success at _k_ measures are well-suited in case there is typically only one relevant document (or retrieving one relevant document is enough).\n",
    "\n",
    " __Implement _Success at 5_ below.__ \n",
    " > The correct result is 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def success_at_5(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    flag = 0\n",
    "    for i in range(5):\n",
    "        if len(retrieved) > 0 and retrieved[i] in relevant:\n",
    "            return 1\n",
    "    return flag\n",
    "    # END ANSWER\n",
    "    \n",
    "success_at_5(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly __implement success at rank 10__\n",
    "\n",
    "> The correct result is 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def success_at_10(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    flag = 0\n",
    "    for i in range(10):\n",
    "        if len(retrieved) > 0 and retrieved[i] in relevant:\n",
    "            return 1\n",
    "    return flag\n",
    "    # END ANSWER\n",
    "    \n",
    "success_at_10(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.A: _Precision, Recall and F-measure_\n",
    "__1. Implement _Precision_ using Formula 8.1 from [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__\n",
    "\n",
    ">_Hint:_ one can count the number of documents in a list using the built-in Python function [len()](https://docs.python.org/3/library/functions.html#len) \\\n",
    "> _example:_ `len(retrieved)` for the number of retrieved documents. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.26666666666666666"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def precision(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    count = 0\n",
    "    for i in retrieved:\n",
    "        if i in relevant:\n",
    "            count += 1\n",
    "    return count/len(retrieved)\n",
    "    # END ANSWER\n",
    "    \n",
    "precision(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Implement _Recall_ using Formula 8.2 from [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4444444444444444"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def recall(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    return len(relevant.intersection(set(retrieved)))/len(relevant)\n",
    "    # END ANSWER\n",
    "    \n",
    "recall(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__3. Implement the balanced F measure (_F_ with β=1) using Formula 8.6 from [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__\n",
    "\n",
    "> Tip: you may reuse your implementations of precision and recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.33333333333333337"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def f_measure(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    return 2*precision(relevant, retrieved)*recall(relevant, retrieved)/(precision(relevant, retrieved)+recall(relevant, retrieved))\n",
    "    # END ANSWER\n",
    "    \n",
    "f_measure(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.B: _Precision at rank k_ and  _R-Precision_\n",
    "\n",
    "Precision, Recall and F are _set_-based measures and suited for unranked lists of documents. If our search system returns a ranked _list_ of results, we can measure precision for several cut-off levels _k_ in the ranked list, i.e. we evaluate the relevance of the TOP-_k_ retrieved documents _(see lecture 01 slides and the book chapter)_. \n",
    "\n",
    "\n",
    "**1. Implement the function `precision_at_k()` that measures the precision at rank _k_**\n",
    "\n",
    "> Interesting fact: For _k_=1, the _Precision at rank 1_ would be the samen as _Success at rank 1_ (why?) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pr@1: 0.00\n",
      "Pr@5: 0.20\n",
      "Pr@10: 0.40\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def precision_at_k(relevant, retrieved, k):\n",
    "    # BEGIN ANSWER\n",
    "    count = 0\n",
    "    if k > 0:\n",
    "        for i in range(k):\n",
    "            if len(retrieved) >= k and retrieved[i] in relevant:\n",
    "                count += 1\n",
    "        return count/k\n",
    "    else:\n",
    "        return 0\n",
    "    # END ANSWER\n",
    "\n",
    "print('Pr@1: %1.2f' % precision_at_k(relevant, retrieved, k=1))\n",
    "print('Pr@5: %1.2f' % precision_at_k(relevant, retrieved, k=5))\n",
    "print('Pr@10: %1.2f' % precision_at_k(relevant, retrieved, k=10))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Implement R-Precision as defined in Chapter 8 of [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book)__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4444444444444444"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def r_precision(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    count = 0\n",
    "    for i in retrieved:\n",
    "        if i in relevant:\n",
    "            count += 1\n",
    "    return count/len(relevant)\n",
    "    # END ANSWER\n",
    "    \n",
    "r_precision(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.D:  Interpolated precision at _recall_ X\n",
    "\n",
    "Another way to address ranked retrieval is to measure precision for several _recall_ levels _X_.\n",
    "\n",
    "__Implement the function `interpolated_precision_at_recall_X()` that measures the interpolated precision at recall level _X_ as defined by formula 8.7 of [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book).__\n",
    "\n",
    "> Tip: calculate for each rank the recall. If the recall is greater than or equal to X, \n",
    "> calculate the precision. Keep the highest (maximum) precision of those to be returned at the end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pr_i@Re01: 0.33\n",
      "Pr_i@Re02: 0.33\n",
      "Pr_i@Re03: 0.33\n",
      "Pr_i@Re04: 0.00\n",
      "Pr_i@Re05: 0.00\n",
      "Pr_i@Re06: 0.00\n",
      "Pr_i@Re07: 0.00\n",
      "Pr_i@Re08: 0.00\n",
      "Pr_i@Re09: 0.00\n",
      "Pr_i@Re10: 0.00\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "def interpolated_precision_at_recall_X (relevant, retrieved, X):\n",
    "    # BEGIN \n",
    "    max_precision = 0\n",
    "    for i in range(1, len(relevant)+1):\n",
    "        # calculate the recall for each item retrieved\n",
    "        recall_val = recall(relevant, retrieved[:i])\n",
    "        if recall_val >= X:\n",
    "            # look to the right of the graph to find the max precision\n",
    "            max_precision = max(max_precision, precision_at_k(relevant, retrieved, i))\n",
    "    return max_precision\n",
    "    # END ANSWER\n",
    "    \n",
    " \n",
    "\n",
    "print('Pr_i@Re01: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.1))\n",
    "print('Pr_i@Re02: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.2))\n",
    "print('Pr_i@Re03: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.3))\n",
    "print('Pr_i@Re04: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.4))\n",
    "print('Pr_i@Re05: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.5))\n",
    "print('Pr_i@Re06: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.6))\n",
    "print('Pr_i@Re07: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.7))\n",
    "print('Pr_i@Re08: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.8))\n",
    "print('Pr_i@Re09: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=0.9))\n",
    "print('Pr_i@Re10: %1.2f' % interpolated_precision_at_recall_X(relevant, retrieved, X=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.E:  _Average Precision_\n",
    "\n",
    "For a single information need, _Average Precision_ is the average of the precision value obtained for the set of top k documents existing after each relevant document is retrieved (see [Manning, Raghavan and Schütze](http://nlp.stanford.edu/IR-book), Pages 159 and 160). \n",
    "\n",
    "__Implement _Average Precision_ for a single information need.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.2206349206349206"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def average_precision(relevant, retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    sum_precision_at_k = 0\n",
    "    for i in range(1, len(relevant)+1): \n",
    "        sum_precision_at_k += precision_at_k(relevant, retrieved, i)\n",
    "    return 1/len(relevant) * sum_precision_at_k \n",
    "    # END ANSWER\n",
    "\n",
    "average_precision(relevant, retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Performance measures in TREC benchmarks\n",
    "\n",
    "The relevance judgments are provided by TREC in so-called _\"qrels\"_ files that look as follows:\n",
    "\n",
    "    1000 Q0 1341 1\n",
    "    1000 Q0 1231 0\n",
    "    1001 Q0 12332 1\n",
    "     ...\n",
    "\n",
    "The columns of the _qrels_ file contain:\n",
    "1. the query identifier\n",
    "2. the query number within that topic (currently unused and should always be Q0)\n",
    "3. the document identifier that was examined by the judges\n",
    "4. the relevance of the document (_1_:relevant; _0_: not relevant).\n",
    "\n",
    "Below we provide some Python code that reads the _qrels_ and the _run_. The qrels will be put in the Python dictionary `all_relevant`. A [Python dictionary](https://docs.python.org/3/tutorial/datastructures.html#dictionaries) provides quick lookup of a set of values given a key. We will use the `query_id` as a key, and a [Python set](https://docs.python.org/3/tutorial/datastructures.html#sets) of relevant document identifiers. For the partial qrels file above, `all_relevant` would look as follows:\n",
    "\n",
    "    {\n",
    "        \"1000\": set([\"1341\", \"1231\"]),\n",
    "        \"1001\": set([\"12332\"])\n",
    "    }\n",
    "    \n",
    "We will use a dictionary called `all_retrieved` with `query_id` as key, and as value a [Python list](https://docs.python.org/3/tutorial/introduction.html#lists) of document identifiers retrieved by the IR system:\n",
    "\n",
    "    {\n",
    "        \"1000\": [\"1341\", \"12346, \"2345\"],\n",
    "        \"1001\": [..., ..., ...],\n",
    "        ...\n",
    "    }\n",
    "\n",
    "Note that, with this data structure, for each `query_id` we can easily access the list of retrieved and relevant documents, and compute the performance metrics. We can then average these measures over all the queries to compute the mean performance of the IR system on the given retrieval task.\n",
    "\n",
    "Please examine the code below, and make sure you understand every line. Use the Python documentation where needed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DATA: the TREC genomics benchmark\n",
    "\n",
    "For the following exercises, we will use a subset of the TREC genomics document collection and queries. \n",
    "It is stored in the folder `data01/` in the directory where you have been instructed to place the assignment notebooks (`/`).\n",
    "\n",
    "The collections contains:\n",
    "\n",
    "* `FIR-s05-medline.json` (the collection in Elasticsearch batch format - because of its size it cannot be indexed with a single curl command!)\n",
    "* `FIR-s05-training-queries-simple.txt` (test queries)\n",
    "* `FIR-s05-training-qrels.txt` (the \"relevance judgements\" for the test queries, i.e. the correct answers)\n",
    "\n",
    "> ___Note___ that these files contain a subset of the documents and queries of the TREC genomics track benchmark, to facilitate experimentations with less computation time needed.\n",
    "> The original files are also included in the `data01/` directory, withouth the `FIR-s05-` prefix (you may use them for the final project).\n",
    "\n",
    "To make things easy, the data is already provided in Elasticsearch' batch processing format. \n",
    "Inspect the collection file in the terminal:\n",
    "\n",
    "`head FIR-s05-medline.json`\n",
    "\n",
    "This shows the first 5 documents in the collection (in JSON format prepared for ElasticSearch, as you have seen in the tutorial)\n",
    "\n",
    "#### Baseline model and results\n",
    "We also provide the list of retrieved documents by a _baseline_ model, in the file `data01/baseline.run`. For each query, it contains the list of document IDs of the retrieved documents (to be compared with those in the qrels file). We use this file in the examples and evaluation exercises below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_qrels_file(qrels_file):  # reads the content of he qrels file\n",
    "    trec_relevant = dict()  # query_id -> set([docid1, docid2, ...])\n",
    "    with open(qrels_file, 'r') as qrels:\n",
    "        for line in qrels:\n",
    "            (qid, q0, doc_id, rel) = line.strip().split()\n",
    "            if qid not in trec_relevant:\n",
    "                # creating a set for each qid, the value of the dictionary is a set\n",
    "                trec_relevant[qid] = set()\n",
    "            if (rel == \"1\"):\n",
    "                trec_relevant[qid].add(doc_id)\n",
    "    return trec_relevant\n",
    "\n",
    "def read_run_file(run_file):  \n",
    "    # read the content of the run file produced by our IR system \n",
    "    # (in the following exercises you will create your own run_files)\n",
    "    trec_retrieved = dict()  # query_id -> [docid1, docid2, ...]\n",
    "    with open(run_file, 'r') as run:\n",
    "        for line in run:\n",
    "            (qid, q0, doc_id, rank, score, tag) = line.strip().split()\n",
    "            if qid not in trec_retrieved:\n",
    "                trec_retrieved[qid] = []\n",
    "            trec_retrieved[qid].append(doc_id) \n",
    "    return trec_retrieved\n",
    "    \n",
    "\n",
    "def read_eval_files(qrels_file, run_file):\n",
    "    return read_qrels_file(qrels_file), read_run_file(run_file)\n",
    "\n",
    "(all_relevant, all_retrieved) = read_eval_files('data01/FIR-s05-training-qrels.txt', 'data01/baseline.run')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### _Number of queries_ and _number of retrieved documents per query_\n",
    " \n",
    "The following code counts the number of queries evaluated in the file `baseline.run` (provided in the `data01/` folder, containing the list of doc-ids retrieved using a baseline model) and prints it (use the result from the cell above). For each query, it also prints the number of documents that were retrieved for that query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of retrieved documents: 38\n",
      "Docs retrieved for query #1: 1000\n",
      "Docs retrieved for query #3: 1000\n",
      "Docs retrieved for query #4: 1000\n",
      "Docs retrieved for query #5: 1000\n",
      "Docs retrieved for query #6: 1000\n",
      "Docs retrieved for query #7: 1000\n",
      "Docs retrieved for query #8: 1000\n",
      "Docs retrieved for query #9: 1000\n",
      "Docs retrieved for query #10: 1000\n",
      "Docs retrieved for query #11: 1000\n",
      "Docs retrieved for query #12: 1000\n",
      "Docs retrieved for query #13: 1000\n",
      "Docs retrieved for query #14: 1000\n",
      "Docs retrieved for query #15: 1000\n",
      "Docs retrieved for query #16: 1000\n",
      "Docs retrieved for query #18: 1000\n",
      "Docs retrieved for query #20: 1000\n",
      "Docs retrieved for query #22: 1000\n",
      "Docs retrieved for query #23: 1000\n",
      "Docs retrieved for query #24: 1000\n",
      "Docs retrieved for query #25: 1000\n",
      "Docs retrieved for query #27: 1000\n",
      "Docs retrieved for query #28: 1000\n",
      "Docs retrieved for query #29: 1000\n",
      "Docs retrieved for query #31: 1000\n",
      "Docs retrieved for query #32: 1000\n",
      "Docs retrieved for query #34: 1000\n",
      "Docs retrieved for query #36: 1000\n",
      "Docs retrieved for query #37: 1000\n",
      "Docs retrieved for query #38: 1000\n",
      "Docs retrieved for query #39: 1000\n",
      "Docs retrieved for query #40: 1000\n",
      "Docs retrieved for query #42: 1000\n",
      "Docs retrieved for query #44: 1000\n",
      "Docs retrieved for query #45: 1000\n",
      "Docs retrieved for query #46: 1000\n",
      "Docs retrieved for query #48: 1000\n",
      "Docs retrieved for query #50: 1000\n"
     ]
    }
   ],
   "source": [
    "print('Number of retrieved documents: %d' % len(all_retrieved))\n",
    "\n",
    "# each qid has 1000 documents, there are 38 unique qid\n",
    "for qid in all_retrieved:\n",
    "    print ('Docs retrieved for query #{}: {}'.format(qid, str(len(all_retrieved[qid]))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For your own understanding, __inspect the structure and content of the `all_retrieved` and `all_relevant` data structures__ to understand them better. Use the `print()` function to see the content of the data structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write here the code to inspect the data structures\n",
    "#print(all_relevant)\n",
    "#print(all_retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 01.F: _mean average precision_\n",
    "__Using the `average_precision()` function you implemented above, write the code to compute the _Mean Average Precision_ for the `baseline.run` results.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean Average Precision (MAP): 0.161\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def mean_average_precision(all_relevant, all_retrieved):\n",
    "    # BEGIN ANSWER\n",
    "    all_relevant_set = set()\n",
    "    all_retrieved_list = []\n",
    "    for key, value in all_retrieved.items():\n",
    "        for i in value:\n",
    "            all_retrieved_list.append(i)\n",
    "    for key, value in all_relevant.items():\n",
    "        for j in value:\n",
    "            all_relevant_set.add(j)\n",
    "            \n",
    "    sum = 0\n",
    "    for k in range(1, len(all_relevant_set)+1):\n",
    "        sum += average_precision(all_relevant_set, all_retrieved_list)\n",
    "    return sum/len(all_relevant_set)\n",
    "    # END ANSWER\n",
    "\n",
    "mapr = mean_average_precision(all_relevant, all_retrieved)\n",
    "print('Mean Average Precision (MAP): %1.3f\\n' % mapr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## TREC benchmark evaluation\n",
    "\n",
    "Below you find a function that take `all_relevant` and `all_retrieved` to compute the mean value of the `measure` over all queries. \n",
    "\n",
    "The function `mean_metric()`'s first function argument, `measure`, is a special argument: it is a function too! The `mean_metric` function sums the total score for the particular measure and divides it by the number of queries. It computes the average measures over all the query results.\n",
    "\n",
    "_This part will be reused later to compare the results of different models._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('mean r_precision', 0.7404266715175034)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def mean_metric(measure, all_relevant, all_retrieved):\n",
    "    total = 0\n",
    "    count = 0\n",
    "    for qid in all_relevant:\n",
    "        relevant  = all_relevant[qid]\n",
    "        retrieved = all_retrieved.get(qid, [])\n",
    "        #print(retrieved)\n",
    "        value = measure(relevant, retrieved)\n",
    "        total += value\n",
    "        count += 1\n",
    "    return \"mean \" + measure.__name__, total / count\n",
    "\n",
    "# Example of use of the mean_metric function: computing the average r_precision\n",
    "mean_metric(r_precision, all_relevant, all_retrieved)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TREC overview of the results of a run\n",
    "The following two functions use your implementation of the metrics to create an overview of the performance metrics on the TREC benchmark data. Give a look at the numbers and make your own interpretations of the results. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for data01/baseline.run\n",
      "mean success_at_1              0.1053\n",
      "mean success_at_5              0.2632\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.7404\n",
      "mean precision_at_1            0.1053\n",
      "mean precision_at_5            0.07895\n",
      "mean precision_at_10           0.04737\n",
      "mean precision_at_50           0.01947\n",
      "mean precision_at_100          0.01395\n",
      "mean precision_at_recall_00    0.1645\n",
      "mean precision_at_recall_01    0.1528\n",
      "mean precision_at_recall_02    0.125\n",
      "mean precision_at_recall_03    0.08114\n",
      "mean precision_at_recall_04    0.07237\n",
      "mean precision_at_recall_05    0.07237\n",
      "mean precision_at_recall_06    0.04605\n",
      "mean precision_at_recall_07    0.04605\n",
      "mean precision_at_recall_08    0.02632\n",
      "mean precision_at_recall_09    0.02632\n",
      "mean precision_at_recall_10    0.02632\n",
      "mean average_precision         0.1012\n"
     ]
    }
   ],
   "source": [
    "def trec_eval(qrels_file, run_file):\n",
    "\n",
    "    def precision_at_1(rel, ret): return precision_at_k(rel, ret, k=1)\n",
    "    def precision_at_5(rel, ret): return precision_at_k(rel, ret, k=5)\n",
    "    def precision_at_10(rel, ret): return precision_at_k(rel, ret, k=10)\n",
    "    def precision_at_50(rel, ret): return precision_at_k(rel, ret, k=50)\n",
    "    def precision_at_100(rel, ret): return precision_at_k(rel, ret, k=100)\n",
    "    def precision_at_recall_00(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.0)\n",
    "    def precision_at_recall_01(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.1)\n",
    "    def precision_at_recall_02(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.2)\n",
    "    def precision_at_recall_03(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.3)\n",
    "    def precision_at_recall_04(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.4)\n",
    "    def precision_at_recall_05(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.5)\n",
    "    def precision_at_recall_06(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.6)\n",
    "    def precision_at_recall_07(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.7)\n",
    "    def precision_at_recall_08(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.8)\n",
    "    def precision_at_recall_09(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=0.9)\n",
    "    def precision_at_recall_10(rel, ret): return interpolated_precision_at_recall_X(rel, ret, X=1.0)\n",
    "\n",
    "    (all_relevant, all_retrieved) = read_eval_files(qrels_file, run_file)\n",
    "    unknown_qids = set(all_retrieved.keys()).difference(all_relevant.keys())\n",
    "    if len(unknown_qids) > 0:\n",
    "        raise ValueError(\"Unknown qids in run: {}\".format(sorted(list(unknown_qids))))\n",
    "\n",
    "    metrics = [success_at_1,\n",
    "               success_at_5,\n",
    "               success_at_10,\n",
    "               r_precision,\n",
    "               precision_at_1,\n",
    "               precision_at_5,\n",
    "               precision_at_10,\n",
    "               precision_at_50,\n",
    "               precision_at_100,\n",
    "               precision_at_recall_00,\n",
    "               precision_at_recall_01,\n",
    "               precision_at_recall_02,\n",
    "               precision_at_recall_03,\n",
    "               precision_at_recall_04,\n",
    "               precision_at_recall_05,\n",
    "               precision_at_recall_06,\n",
    "               precision_at_recall_07,\n",
    "               precision_at_recall_08,\n",
    "               precision_at_recall_09,\n",
    "               precision_at_recall_10,\n",
    "               average_precision]\n",
    "\n",
    "    return [mean_metric(metric, all_relevant, all_retrieved) for metric in metrics]\n",
    "\n",
    "\n",
    "def print_trec_eval(qrels_file, run_file):\n",
    "    results = trec_eval(qrels_file, run_file)\n",
    "    print(\"Results for {}\".format(run_file))\n",
    "    for (metric, score) in results:\n",
    "        print(\"{:<30} {:.4}\".format(metric, score))\n",
    "\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'data01/baseline.run')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Exercise 01.H: _Significance testing_\n",
    "\n",
    "Testing the statistical significance of differences of the results of different IR systems is important (see slides of lecture 01 and course book, Section 8.8). One of the basic tests one can perform is the two-tailed [sign test](https://en.wikipedia.org/wiki/Sign_test).\n",
    "\n",
    "Only for this exercise, we use the run files obtained by  [Hiemstra and Aly](https://djoerdhiemstra.com/wp-content/uploads/trec2014mirex-draft.pdf) for the TREC Web track 2014 benchmark (note these files are from a different benchmark from what we have been working with so far). The `utbase.run` file was generated using Language Modeling, while `utexact.run` was generated using an IR system based on mathing the exact query string, and ranking the documents by  the number of exact matches found. The exact run improves the _Precision at 5_ to 0.456 (compared to 0.440 for the baseline run).  \n",
    "\n",
    "__Implement the code to perform the _sign test_ of statistical significance.__\n",
    "> _Hint:_ for each sign, compute the number of queries that increase/descrease performance (called `better, worse` in the code below). How would you use these values to compute the _p_ value of the two-tailed sign test? Is the difference between _utbase_ and _utexact_ significant?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 0)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def sign_test_values(measure, qrels_file, run_file_1, run_file_2):\n",
    "    all_relevant = read_qrels_file(qrels_file)\n",
    "    all_retrieved_1 = read_run_file(run_file_1)\n",
    "    all_retrieved_2 = read_run_file(run_file_2)\n",
    "    better = 0\n",
    "    worse  = 0\n",
    "    # BEGIN ANSWER\n",
    "    # END ANSWER\n",
    "    return(better, worse)\n",
    "    \n",
    "def precision_at_rank_5(rel, ret):\n",
    "    return precision_at_k(rel, ret, k=5)\n",
    "\n",
    "sign_test_values(precision_at_rank_5, 'data01/trec.qrels', 'data01/utbase.run', 'data01/utexact.run')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BEGIN ANSWER\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "***\n",
    "***\n",
    "***\n",
    "\n",
    "# Part 02 - Indexing and querying with ElasticSearch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation: Getting started with Elasticsearch\n",
    "\n",
    "The following parts of the assignment will be based on ElasticSearch. you are adviced to go through the \"Elasticsearch, [reference guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/getting-started.html)\", and work on the tutorials. You can skip the section on [Installation](https://www.elastic.co/guide/en/elasticsearch/reference/current/_installation.html), as we provide it already installed in the Virtual Machine.\n",
    "\n",
    "> If you want (disclaimer: we do __not__ give help with this!), you can \n",
    "> follow the [Installation](https://www.elastic.co/guide/en/elasticsearch/reference/current/_installation.html) to run Elasticsearch on your laptop without VM. Beware your system will likely be different from the \n",
    "> one of your colleagues and they might not be able to help you if \n",
    "> you have problems that are specific to your system, your operating\n",
    "> system, or your Elasticsearch version.\n",
    "\n",
    "### Starting/Stopping ElasticSearch\n",
    "To start ElasticSearch on the virtual machine, you can type `sudo service elasticsearch start` in a Terminal.\n",
    "To stop the ElasticSearch server, instead, you can type `sudo service elasticsearch stop`. Refer at the [the official guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/deb.html#deb-running-init), for more information.\n",
    "\n",
    "### The REST API\n",
    "\n",
    "Elasticsearch runs its own server that can be accessed by a regular web browser by opening this link: http://localhost:9200. \n",
    "\n",
    "Elasticsearch will respond with something like:\n",
    "\n",
    "    {\n",
    "        \"name\" : \"fir-machine\",\n",
    "        \"cluster_name\" : \"elasticsearch\",\n",
    "        \"cluster_uuid\" : \"w7SBVo1ESVivMApbLIqRvA\",\n",
    "        \"version\" : {\n",
    "            \"number\" : \"7.9.0\",\n",
    "            \"build_flavor\" : \"default\",\n",
    "            \"build_type\" : \"deb\",\n",
    "            \"build_hash\" : \"a479a2a7fce0389512d6a9361301708b92dff667\",\n",
    "            \"build_date\" : \"2020-08-11T21:36:48.204330Z\",\n",
    "            \"build_snapshot\" : false,\n",
    "            \"lucene_version\" : \"8.6.0\",\n",
    "            \"minimum_wire_compatibility_version\" : \"6.8.0\",\n",
    "            \"minimum_index_compatibility_version\" : \"6.0.0-beta1\"\n",
    "        },\n",
    "        \"tagline\" : \"You Know, for Search\"\n",
    "    }\n",
    "\n",
    "\n",
    "If you see this, then your Elasticsearch node is up and running. The RESTful API uses simple text or JSON over HTTP. \n",
    "\n",
    "> REST, API, JSON, HTTP, that's a lot of abbreviations! It is good to\n",
    "> be familiar with the terminology. Let us explain: The Elasticsearch\n",
    "> response is not (only) intended for humans. It is supposed to be used \n",
    "> by applications that run on the client machines, and therefore the\n",
    "> interface is called an Application Programming Interface (API). The \n",
    "> API uses a format called JSON (JavaScript Object Notation), which \n",
    "> can be easily read by machines (and humans). The API sends its JSON\n",
    "> response using the same method as your web browser displays web\n",
    "> pages. This method is called HTTP (Hyper Text Transfer Protocol), \n",
    "> and it is the reason you can inspect the response in a normal web\n",
    "> browser. APIs that use HTTP are called RESTful interfaces. REST \n",
    "> stands for REpresentational State Transfer, arguably one of the\n",
    "> simplest ways to define an API.\n",
    "\n",
    "\n",
    "### Interacting with the ElasticSearch server\n",
    "\n",
    "You can interact with your Elasticsearch service in different ways. In this first part we explore Kibana, a dashboard for inspection of your indices. Later during the practical work we will use the Python Elasticsearch client or the DSL library. You can also start yourself with Python.\n",
    "\n",
    "#### Kibana\n",
    "Kibana provides a web interface to interact with your Elasticsearch service. It's available from http://localhost:5601. You can use Kibana to create interactive dashboards visualizing data in your Elasticsearch indices. It also provides a console to execute Elasticsearch commands. It's available from http://localhost:5601/app/kibana#/dev_tools\n",
    "\n",
    "To start Kibana on the virtual machine, you can type `sudo service kibana start` in a Terminal. \\\n",
    "To stop the Kibana server, instead, you can type `sudo service kibana stop`.\n",
    "\n",
    "Many examples from the Elasticsearch user guide can be directly executed in Kibana by clicking on the `CONSOLE` button.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "***\n",
    "***\n",
    "\n",
    "# Indexing and queries (Exercises - Part 02)\n",
    "\n",
    "_You can work on this part after Lecture 01 already_\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collection indexing: useful code\n",
    "\n",
    "We provide some code to read the TREC collection documents and index them into the ElasticSearch engine.\n",
    "As we need to re-index the document collection when we use a different indexing configurations (called Mappings in ElasticSearch), we developed some functions to support a quick re-indexing in the following exercises.\n",
    "\n",
    "Below you find the Python code for bulk-indexing our (FIR)Medline collection. Execute the following cells to index the collection in an Elasticsearch index called `genomics'. Study the code carefully, as you will use the indexing functions later for the completion of the assignment.\n",
    "\n",
    "> The code uses additional helper functions \n",
    "> (`elasticsearch.helpers`) and a library for processing JSON.\n",
    "> The function `read_documents()` reads the bulk collection file: The \n",
    "> function is a [Python generator](https://wiki.python.org/moin/Generators) function. It generates an 'on-demand' list\n",
    "> by using the statement `yield` for every item of the list. It\n",
    "> is used in the helper function `elasticsearch.helpers.bulk()`.\n",
    "> The statement `raise` is Python's approach to throw exceptions: it exits the program with an error.\n",
    "> Note the (keyword) arguments to bulk:\n",
    "> `chunk_size` indicates the number of documents to be processed by\n",
    "> elasticsearch in one batch. \n",
    "> The request_timeout is set to 30 seconds because processing a single batch\n",
    "> of documents can take some time.\n",
    "\n",
    "> __Note:__ _when processing a bulk index, be sure to have few GigaBytes free on the hard drive of the VM. If you get a BulkIndexError with read-only/FORBIDDEN errors, you probably have too little hard drive space available for ElasticSearch to work properly._\n",
    "\n",
    "\n",
    "**_Note:_ indexing the (FIR)TREC genomics collection can take some time, be patient.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import elasticsearch\n",
    "import elasticsearch.helpers\n",
    "import json\n",
    "\n",
    "def read_documents(file_name):\n",
    "    \"\"\"\n",
    "    Returns a generator of documents to be indexed by elastic, read from file_name\n",
    "    \"\"\"\n",
    "    with open(file_name, 'r') as documents:\n",
    "        for line in documents:\n",
    "            doc_line = json.loads(line)\n",
    "            if ('index' in doc_line):\n",
    "                id = doc_line['index']['_id']\n",
    "            elif ('PMID' in doc_line):\n",
    "                doc_line['_id'] = id\n",
    "                yield doc_line\n",
    "            else:\n",
    "                raise ValueError('Woops, error in index file')\n",
    "\n",
    "def create_index(es, index_name, body={}):\n",
    "    # delete index when it already exists\n",
    "    es.indices.delete(index=index_name, ignore=[400, 404])\n",
    "    # create the index \n",
    "    es.indices.create(index=index_name, body=body)\n",
    "                \n",
    "def index_documents(es, collection_file_name, index_name, body={}):\n",
    "    create_index(es, index_name, body)\n",
    "    # bulk index the documents from file_name\n",
    "    return elasticsearch.helpers.bulk(\n",
    "        es, \n",
    "        read_documents(collection_file_name),\n",
    "        index=index_name,\n",
    "        chunk_size=2000,\n",
    "        request_timeout=30\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'index_documents' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [21], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Index the collection into the index called 'genomics'\u001b[39;00m\n\u001b[1;32m      5\u001b[0m body \u001b[38;5;241m=\u001b[39m {} \u001b[38;5;66;03m# no indexing options (leave default)\u001b[39;00m\n\u001b[0;32m----> 6\u001b[0m \u001b[43mindex_documents\u001b[49m(es, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdata01/FIR-s05-medline.json\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenomics-base\u001b[39m\u001b[38;5;124m'\u001b[39m, body)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'index_documents' is not defined"
     ]
    }
   ],
   "source": [
    "# Connect to the ElasticSearch server\n",
    "es = elasticsearch.Elasticsearch(host='localhost')  # in case you use Docker, the host is 'elasticsearch'\n",
    "\n",
    "# Index the collection into the index called 'genomics'\n",
    "body = {} # no indexing options (leave default)\n",
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-base', body)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> You can change the name of the index, in case you want to have different indices of the same collection created with different indexing settings, and compare the performance on the test queries. \n",
    "\n",
    "> E.g. you create two indices 'genomics01' and 'genomics02': genomics01 uses the default options, while genomics02 uses custom tokenizers. You will then have two indices with different characteristics (and probably different performance). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 02.A: index properties and querying"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__1. Query the index called 'genomics-base' and determine how many documents are indexed.__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prevent having to index the collection again\n",
    "import elasticsearch\n",
    "import elasticsearch.helpers\n",
    "import json\n",
    "\n",
    "es = elasticsearch.Elasticsearch(host='localhost')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "263080\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code here\n",
    "# BEGIN ANSWER\n",
    "index_name = 'genomics-base'\n",
    "resp = es.count(index=index_name)\n",
    "print(resp['count'])\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. How many documents containing the term `molecule` are there in your index? (searching all fields of the documents).__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3404\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code that generates the answer here (you may also use Kibana)\n",
    "# BEGIN ANSWER\n",
    "from elasticsearch_dsl import Search, Q   \n",
    "\n",
    "s = Search(using=es, index=index_name).query(\"multi_match\", query=\"molecule\", fields=[\"*\"])\n",
    "print(s.count())\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__3. How many documents containing the term `molecular` are there in your index? (searching all fields of the documents).__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31556\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code that generates the answer here (you may also use Kibana)\n",
    "# BEGIN ANSWER\n",
    "s = Search(using=es, index=index_name).query(\"multi_match\", query=\"molecular\", fields=[\"*\"])\n",
    "print(s.count())\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__4. How many documents containing the terms `cell` AND `blood` are there in your index? (searching all fields of the documents).__\n",
    "\n",
    "You can use Kibana (suggested for the time being - you can use the command line in Kibana), the Python ElasticSearch library or DSL. Report the code you implemented and the resulting number of documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6865\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write the code that generates the answer here (you may also use Kibana)\n",
    "# BEGIN ANSWER\n",
    "\n",
    "# chaining query method, it implicitly calls the & operator\n",
    "s = Search(using=es, index=index_name).query(\"multi_match\", query=\"cell\").query(\"multi_match\", query=\"blood\")\n",
    "print(s.count())\n",
    "\n",
    "# using Q\n",
    "# s1 = Search(using=es)\n",
    "# q = Q(\"multi_match\", query=\"cell\") & Q(\"multi_match\", query=\"blood\")\n",
    "# s1 = s1.query(q)\n",
    "# print(s1.count())\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of results: 6865\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_364/1364412639.py:10: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  result = es.search(index='genomics-base', body=body)\n"
     ]
    }
   ],
   "source": [
    "import elasticsearch\n",
    "es = elasticsearch.Elasticsearch(host='localhost')  # in case you use Docker, the host is 'elasticsearch'\n",
    "\n",
    "# this is another solution (if query_string is used, be sure that AND is in the query, otherwise it will not search properly)\n",
    "term = 'blood AND cell'\n",
    "body = {\"track_total_hits\": True, \"query\": {\"query_string\": \n",
    "                                            {\"query\": term, \n",
    "                                             \"default_operator\":\"AND\", \n",
    "                                             \"auto_generate_synonyms_phrase_query\": True }}}\n",
    "result = es.search(index='genomics-base', body=body)\n",
    "print(\"Number of results: {}\".format(result['hits']['total']['value']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 02.B: the Python ElasticSearch library\n",
    "\n",
    "#### Preparation\n",
    "The command line is fine for doing basic operations on your Elasticsearch indices, but as soon as things get more complex, you better use custom client programs.\n",
    "We will use the [Elasticsearch client library for Python](https://elasticsearch-py.readthedocs.io). This library will execute the HTTP requests that you have used before (with CURL or Kibana). The library is pre-installed on the VM.\n",
    "\n",
    "#### Exercise\n",
    "\n",
    "__Write the code that searches the index for _\"molecule\"_ using the [search()](https://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.Elasticsearch.search) function.__ Your code will take at minimum the following steps:\n",
    "\n",
    "1. import the python library `elasticsearch`.\n",
    "2. open a connection with the Elasticsearch host `'elasticsearch'` with `Elasticsearch()`.\n",
    "3. execute a search with `search()` using the index `genomics-base`, and a correct query body.\n",
    "4. print the JSON output of Elasticsearch \n",
    "\n",
    "How many hits are there in your index? Is the result the same as in Exercise 02.A?\n",
    "\n",
    "> Elasticsearch runs on localhost on your laptop, at port 9200 (so as http://localhost:9200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'_shards': {'failed': 0, 'skipped': 0, 'successful': 1, 'total': 1},\n",
      " 'hits': {'hits': [{'_id': '374140',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 31.376125,\n",
      "                    '_source': {'AB': 'OBJECTIVE: Endothelial cell dysfunction '\n",
      "                                      'has been implicated in the inflammatory '\n",
      "                                      'response to cardiopulmonary bypass, and '\n",
      "                                      'the upregulation of endothelial cell '\n",
      "                                      'expression of adhesion molecules might '\n",
      "                                      'promote leukocyte extravasation in '\n",
      "                                      'vivo. Soluble endothelial cell adhesion '\n",
      "                                      'molecules are increased after bypass. '\n",
      "                                      'The aim of this study was to '\n",
      "                                      'investigate the relationship between '\n",
      "                                      'endothelial cell-surface expression of '\n",
      "                                      'adhesion molecules and their '\n",
      "                                      'concentration in plasma after coronary '\n",
      "                                      'artery bypass grafting. METHODS: Ten '\n",
      "                                      'patients undergoing coronary artery '\n",
      "                                      'bypass with cardiopulmonary bypass had '\n",
      "                                      '5 plasma samples taken at defined '\n",
      "                                      'intervals before, during, and after '\n",
      "                                      'cardiopulmonary bypass. Plasma was '\n",
      "                                      'incubated with human umbilical vein '\n",
      "                                      'endothelial cell monolayers, and '\n",
      "                                      'expression of E-selectin, intercellular '\n",
      "                                      'adhesion molecule 1, and vascular cell '\n",
      "                                      'adhesion molecule 1 on the surface of '\n",
      "                                      'human umbilical vein endothelial cell '\n",
      "                                      'monolayers was measured by means of '\n",
      "                                      'enzyme-linked immunosorbent assay. '\n",
      "                                      'Plasma soluble adhesion molecules, '\n",
      "                                      'C-reactive protein, interleukin 8, '\n",
      "                                      'interleukin 10, transforming growth '\n",
      "                                      'factor beta1, and neutrophil counts '\n",
      "                                      'were determined for each patient. '\n",
      "                                      'RESULTS: Markers typical of acute '\n",
      "                                      'inflammation (ie, interleukin 8, '\n",
      "                                      'neutrophils, and C-reactive protein) '\n",
      "                                      'were all increased after bypass. '\n",
      "                                      'Soluble plasma intercellular and '\n",
      "                                      'vascular cell adhesion molecule 1 (but '\n",
      "                                      'not E-selectin) were increased after '\n",
      "                                      'bypass. However, endothelial cell '\n",
      "                                      'expression of vascular cell adhesion '\n",
      "                                      'molecule 1 and intercellular adhesion '\n",
      "                                      'molecule 1 (but not E-selectin) were '\n",
      "                                      'significantly decreased by exposure to '\n",
      "                                      'postbypass plasma. Additionally, '\n",
      "                                      'postbypass plasma inhibited interleukin '\n",
      "                                      '1beta-stimulated endothelial cell '\n",
      "                                      'expression of vascular cell and '\n",
      "                                      'intercellular adhesion molecule 1. '\n",
      "                                      'Interleukin 10 and transforming growth '\n",
      "                                      'factor beta1, both of which are known '\n",
      "                                      'to inhibit endothelial cell adhesion '\n",
      "                                      'molecule expression, were respectively '\n",
      "                                      'increased 10-fold and 3-fold (P <.05) '\n",
      "                                      'after bypass. CONCLUSIONS: Despite '\n",
      "                                      'containing increased soluble '\n",
      "                                      'intercellular and vascular cell '\n",
      "                                      'adhesion molecule 1, postbypass plasma '\n",
      "                                      'inhibits endothelial cell expression of '\n",
      "                                      'intercellular and vascular cell '\n",
      "                                      'adhesion molecule 1. Upregulated '\n",
      "                                      'vascular expression of adhesion '\n",
      "                                      'molecules might not be essential for '\n",
      "                                      'endothelial activation after bypass.',\n",
      "                                'AD': 'Clinical Research Group, The Heart '\n",
      "                                      'Research Institute, Cardiothoracic '\n",
      "                                      'Surgical Unit, Royal Prince Alfred '\n",
      "                                      'Hospital, Missenden Road, Camperdown, '\n",
      "                                      'New South Wales 2050, Australia. '\n",
      "                                      'valsby@hotmail.com',\n",
      "                                'AID': 'S0022522302001484 [pii]',\n",
      "                                'CY': 'United States',\n",
      "                                'DA': '20020926',\n",
      "                                'DCOM': '20021209',\n",
      "                                'DP': '2002 Oct',\n",
      "                                'EDAT': '2002/09/27 06:00',\n",
      "                                'FAU': ['Vallely, Michael P',\n",
      "                                        'Bannon, Paul G',\n",
      "                                        'Hughes, Clifford F',\n",
      "                                        'Kritharides, Leonard'],\n",
      "                                'IP': '4',\n",
      "                                'IS': '0022-5223',\n",
      "                                'JID': '0376343',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['*Cardiopulmonary Bypass',\n",
      "                                       'Coronary Arteriosclerosis/surgery',\n",
      "                                       'Coronary Artery Bypass',\n",
      "                                       'E-Selectin/*blood',\n",
      "                                       'Endothelium, Vascular/*metabolism',\n",
      "                                       'Human',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/*blood/metabolism',\n",
      "                                       'Interleukin-1/*blood',\n",
      "                                       'Interleukin-8/*blood',\n",
      "                                       \"Support, Non-U.S. Gov't\",\n",
      "                                       'Up-Regulation',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/*blood/metabolism'],\n",
      "                                'MHDA': '2002/12/10 04:00',\n",
      "                                'PG': '758-67',\n",
      "                                'PMID': '12324734',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (E-Selectin)',\n",
      "                                       '0 (Interleukin-1)',\n",
      "                                       '0 (Interleukin-8)',\n",
      "                                       '0 (Vascular Cell Adhesion Molecule-1)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'J Thorac Cardiovasc Surg 2002 '\n",
      "                                      'Oct;124(4):758-67.',\n",
      "                                'TA': 'J Thorac Cardiovasc Surg',\n",
      "                                'TI': 'Endothelial expression of intercellular '\n",
      "                                      'adhesion molecule 1 and vascular cell '\n",
      "                                      'adhesion molecule 1 is suppressed by '\n",
      "                                      'postbypass plasma containing increased '\n",
      "                                      'soluble intercellular adhesion molecule '\n",
      "                                      '1 and vascular cell adhesion molecule '\n",
      "                                      '1.',\n",
      "                                'UI': '22236900',\n",
      "                                'VI': '124'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '261061',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 28.797413,\n",
      "                    '_source': {'AB': 'The aim of our study was to investigate '\n",
      "                                      'the influence of single low-density '\n",
      "                                      'lipoprotein apheresis (heparin '\n",
      "                                      'extracorporeal low-density lipoprotein '\n",
      "                                      'precipitation [HELP]procedure) on '\n",
      "                                      'plasma concentrations of soluble '\n",
      "                                      'adhesion molecules (sAMs) such as '\n",
      "                                      'soluble vascular cellular adhesion '\n",
      "                                      'molecule-1 (sVCAM-1), soluble '\n",
      "                                      'intercellular adhesion molecule-1 '\n",
      "                                      '(sICAM-1), and P-selectin in patients '\n",
      "                                      'with familial heterozygous '\n",
      "                                      'hypercholesterolemia and documented '\n",
      "                                      'coronary artery disease enrolled in a '\n",
      "                                      'chronic weekly HELP apheresis. Before '\n",
      "                                      'HELP apheresis, the mean plasma '\n",
      "                                      'concentration of sVCAM-1 was 515 +/',\n",
      "                                'AD': 'Department of Cardiology, Heart Center '\n",
      "                                      'North Rhine-Westphalia, Ruhr University '\n",
      "                                      'of Bochum, Bad Oeynhausen, Germany.',\n",
      "                                'AID': 'tap405 [pii]',\n",
      "                                'CY': 'United States',\n",
      "                                'DA': '20020711',\n",
      "                                'DCOM': '20020823',\n",
      "                                'DP': '2002 Jun',\n",
      "                                'EDAT': '2002/07/12 10:00',\n",
      "                                'FAU': ['Pulawski, Ewa',\n",
      "                                        'Mellwig, Klaus-Peter',\n",
      "                                        'Brinkmann, Thomas',\n",
      "                                        'Kleesiek, Knut',\n",
      "                                        'Horstkotte, Dieter'],\n",
      "                                'IP': '3',\n",
      "                                'IS': '1091-6660',\n",
      "                                'JID': '9706703',\n",
      "                                'LA': 'eng',\n",
      "                                'LR': '20021101',\n",
      "                                'MH': ['Adolescent',\n",
      "                                       'Adult',\n",
      "                                       'Aged',\n",
      "                                       'Blood Component Removal/*methods',\n",
      "                                       'Child',\n",
      "                                       'Cholesterol/blood',\n",
      "                                       'Coronary Disease/complications',\n",
      "                                       'Female',\n",
      "                                       'Human',\n",
      "                                       'Hypercholesterolemia, '\n",
      "                                       'Familial/blood/*therapy',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/*blood',\n",
      "                                       'Lipoproteins, LDL/*blood',\n",
      "                                       'Male',\n",
      "                                       'Middle Age',\n",
      "                                       'Selectins/*blood',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/*blood'],\n",
      "                                'MHDA': '2002/08/24 10:01',\n",
      "                                'PG': '229-33',\n",
      "                                'PMID': '12109949',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (Lipoproteins, LDL)',\n",
      "                                       '0 (Selectins)',\n",
      "                                       '0 (Vascular Cell Adhesion Molecule-1)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)',\n",
      "                                       '57-88-5 (Cholesterol)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'Ther Apher 2002 Jun;6(3):229-33.',\n",
      "                                'TA': 'Ther Apher',\n",
      "                                'TI': 'Influence of single low-density '\n",
      "                                      'lipoprotein apheresis on the adhesion '\n",
      "                                      'molecules soluble vascular cellular '\n",
      "                                      'adhesion molecule-1, soluble '\n",
      "                                      'intercellular adhesion molecule-1, and '\n",
      "                                      'P-selectin.',\n",
      "                                'UI': '22107830',\n",
      "                                'VI': '6'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '175209',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 27.891474,\n",
      "                    '_source': {'AB': 'OBJECTIVES : The molecular mechanisms '\n",
      "                                      'whereby severe, uncontrolled '\n",
      "                                      'hypertension (SHT) is translated into '\n",
      "                                      'acute vascular target organ dysfunction '\n",
      "                                      'have not been completely defined. We '\n",
      "                                      'sought to determine whether SHT is '\n",
      "                                      'associated with pressure-dependent '\n",
      "                                      'endothelial activation as assessed by '\n",
      "                                      'soluble vascular cell adhesion '\n",
      "                                      'molecule-1 (sVCAM-1), soluble '\n",
      "                                      'intercellular adhesion molecule-1 '\n",
      "                                      '(sICAM-1) and von Willebrand Factor '\n",
      "                                      '(vWF). METHODS : We determined sVCAM-1, '\n",
      "                                      'sICAM-1 and vWF in three groups: (i) '\n",
      "                                      'untreated patients referred '\n",
      "                                      'specifically for treatment of SHT '\n",
      "                                      '[diastolic blood pressure (DBP) > or = '\n",
      "                                      '120 mm Hg; n = 24]; (ii) untreated '\n",
      "                                      'patients with established mild '\n",
      "                                      'hypertension (MHT; DBP 95-100 mmHg; n = '\n",
      "                                      '19); and (iii) normotensive volunteers '\n",
      "                                      '(DBP < or = 90; n = 16). RESULTS : By '\n",
      "                                      'analysis of variance, sVCAM-1 (P = '\n",
      "                                      '0.002), sICAM-1 (P = 0.02) and vWF (P = '\n",
      "                                      '0.009) were greater in SHT and MHT than '\n",
      "                                      'in normotensives but did not differ '\n",
      "                                      'between SHT and MHT. We observed a '\n",
      "                                      'significant positive correlation '\n",
      "                                      'between blood pressure and soluble '\n",
      "                                      'activation markers at lower blood '\n",
      "                                      'pressures (normotensives and MHT '\n",
      "                                      'considered together) that was not '\n",
      "                                      'present in SHT. CONCLUSIONS : Even mild '\n",
      "                                      'elevation of blood pressure may be '\n",
      "                                      'sufficient to activate the expression '\n",
      "                                      'of adhesion molecules. Mechanisms other '\n",
      "                                      'than the endothelial expression of '\n",
      "                                      'adhesion molecules may be important in '\n",
      "                                      'mediating the accelerated target organ '\n",
      "                                      'injury produced by SHT in humans. '\n",
      "                                      'Concentrations of soluble adhesion '\n",
      "                                      'molecules and vWF may depend more '\n",
      "                                      'strongly upon factors in the '\n",
      "                                      'hypertensive microenvironment other '\n",
      "                                      'than the absolute level of blood '\n",
      "                                      'pressure.',\n",
      "                                'AD': 'Division of Clinical Pharmacology, '\n",
      "                                      'Department of Medicine, University of '\n",
      "                                      'Miami School of Medicine, Miami, '\n",
      "                                      'Florida 33136, USA. '\n",
      "                                      'rpreston@med.miami.edu',\n",
      "                                'CIN': 'J Hypertens. 2002 May;20(5):813-6. '\n",
      "                                       'PMID: 12011632',\n",
      "                                'CY': 'England',\n",
      "                                'DA': '20020515',\n",
      "                                'DCOM': '20020927',\n",
      "                                'DP': '2002 May',\n",
      "                                'EDAT': '2002/05/16 10:00',\n",
      "                                'FAU': ['Preston, Richard A',\n",
      "                                        'Ledford, Marlies',\n",
      "                                        'Materson, Barry J',\n",
      "                                        'Baltodano, Neyton M',\n",
      "                                        'Memon, Abdul',\n",
      "                                        'Alonso, Alberto'],\n",
      "                                'IP': '5',\n",
      "                                'IS': '0263-6352',\n",
      "                                'JID': '8306882',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['Adult',\n",
      "                                       'Blood Pressure',\n",
      "                                       'Comparative Study',\n",
      "                                       'Cross-Sectional Studies',\n",
      "                                       'Diastole',\n",
      "                                       'Endothelium, Vascular/*physiopathology',\n",
      "                                       'Female',\n",
      "                                       'Human',\n",
      "                                       'Hypertension/*physiopathology',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/*metabolism',\n",
      "                                       'Male',\n",
      "                                       'Middle Age',\n",
      "                                       'Reference Values',\n",
      "                                       'Severity of Illness Index',\n",
      "                                       'Solubility',\n",
      "                                       'Systole',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/*metabolism',\n",
      "                                       'von Willebrand Factor/*metabolism'],\n",
      "                                'MHDA': '2002/09/28 04:00',\n",
      "                                'PG': '871-7',\n",
      "                                'PMID': '12011647',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (Vascular Cell Adhesion Molecule-1)',\n",
      "                                       '0 (von Willebrand Factor)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'J Hypertens 2002 May;20(5):871-7.',\n",
      "                                'TA': 'J Hypertens',\n",
      "                                'TI': 'Effects of severe, uncontrolled '\n",
      "                                      'hypertension on endothelial activation: '\n",
      "                                      'soluble vascular cell adhesion '\n",
      "                                      'molecule-1, soluble intercellular '\n",
      "                                      'adhesion molecule-1 and von Willebrand '\n",
      "                                      'factor.',\n",
      "                                'UI': '22007421',\n",
      "                                'VI': '20'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '435276',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 27.877588,\n",
      "                    '_source': {'AB': 'Leptomeningeal metastases occur in up '\n",
      "                                      'to 8% of patients with systemic '\n",
      "                                      'malignancies and have a poor prognosis. '\n",
      "                                      'A better understanding of the '\n",
      "                                      'pathophysiologic processes underlying '\n",
      "                                      'leptomeningeal metastases is needed for '\n",
      "                                      'more effective treatment strategies. We '\n",
      "                                      'hypothesized that tumor cells will have '\n",
      "                                      'to adhere to the well-vascularized '\n",
      "                                      'leptomeninges, because the '\n",
      "                                      'cerebrospinal fluid lacks nutrients and '\n",
      "                                      'growth factors for efficient tumor cell '\n",
      "                                      'proliferation. Specific receptor-ligand '\n",
      "                                      'interactions, which are unknown until '\n",
      "                                      'now, will mediate this adhesion '\n",
      "                                      'process. We determined the growth '\n",
      "                                      'characteristics of B16F-10 melanoma '\n",
      "                                      'cells in cerebrospinal fluid. The '\n",
      "                                      'expression levels of specific adhesion '\n",
      "                                      'molecules on both mouse leptomeningeal '\n",
      "                                      'cells (MLMC) and murine B16F-10 '\n",
      "                                      'melanoma cells were measured by '\n",
      "                                      'immunofluorescence flow cytometry. We '\n",
      "                                      'used mAbs to determine the function of '\n",
      "                                      'these specific adhesion molecules on '\n",
      "                                      'B16F-10 melanoma cell adhesion to a '\n",
      "                                      'leptomeningeal cell layer under static '\n",
      "                                      'and (cerebrospinal fluid-like) flow '\n",
      "                                      'conditions. B16F-10 melanoma cells did '\n",
      "                                      'not proliferate in cerebrospinal fluid '\n",
      "                                      'because of a lack of nutrients and '\n",
      "                                      'growth factors. MLMC expressed low '\n",
      "                                      'levels of vascular cell adhesion '\n",
      "                                      'molecule-1 (VCAM-1), intercellular '\n",
      "                                      'adhesion molecule-1 (ICAM-1), beta1',\n",
      "                                'AD': 'Department of Neurology, University '\n",
      "                                      'Medical Center Utrecht, Utrecht, The '\n",
      "                                      'Netherlands.',\n",
      "                                'CY': 'United States',\n",
      "                                'DA': '20021113',\n",
      "                                'DCOM': '20021210',\n",
      "                                'DP': '2002 Nov',\n",
      "                                'EDAT': '2002/11/14 04:00',\n",
      "                                'FAU': ['Brandsma, Dieta',\n",
      "                                        'Reijneveld, Jaap C',\n",
      "                                        'Taphoorn, Martin J B',\n",
      "                                        'de Boer, Hetty C',\n",
      "                                        'Gebbink, Martijn F B G',\n",
      "                                        'Ulfman, Laurien H',\n",
      "                                        'Zwaginga, Jaap-Jan',\n",
      "                                        'Voest, Emile E'],\n",
      "                                'IP': '11',\n",
      "                                'IS': '0023-6837',\n",
      "                                'JID': '0376617',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['Animal',\n",
      "                                       'Antigens, CD29/analysis',\n",
      "                                       'Antigens, CD44/analysis',\n",
      "                                       'Cell Adhesion',\n",
      "                                       'E-Selectin/analysis',\n",
      "                                       'Human',\n",
      "                                       'Integrin beta3/analysis',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/analysis',\n",
      "                                       'Melanoma, Experimental/*pathology',\n",
      "                                       'Meningeal Neoplasms/*secondary',\n",
      "                                       'Mice',\n",
      "                                       'Stress, Mechanical',\n",
      "                                       \"Support, Non-U.S. Gov't\",\n",
      "                                       'Tumor Necrosis Factor/pharmacology',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/analysis/*physiology'],\n",
      "                                'MHDA': '2002/12/11 04:00',\n",
      "                                'PG': '1493-502',\n",
      "                                'PMID': '12429809',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (Antigens, CD29)',\n",
      "                                       '0 (Antigens, CD44)',\n",
      "                                       '0 (E-Selectin)',\n",
      "                                       '0 (Integrin beta3)',\n",
      "                                       '0 (Tumor Necrosis Factor)',\n",
      "                                       '0 (Vascular Cell Adhesion Molecule-1)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'Lab Invest 2002 Nov;82(11):1493-502.',\n",
      "                                'TA': 'Lab Invest',\n",
      "                                'TI': 'Vascular cell adhesion molecule-1 is a '\n",
      "                                      'key adhesion molecule in melanoma cell '\n",
      "                                      'adhesion to the leptomeninges.',\n",
      "                                'UI': '22317182',\n",
      "                                'VI': '82'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '15327',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 27.811653,\n",
      "                    '_source': {'AB': 'AIMS: To investigate localization and '\n",
      "                                      'concentration of adhesion molecules in '\n",
      "                                      'the human lower uterine segment during '\n",
      "                                      'preterm parturition. METHODS: In a '\n",
      "                                      'prospective study biopsy specimens from '\n",
      "                                      'the lower uterine segment of 43 '\n",
      "                                      'patients undergoing cesarean section at '\n",
      "                                      \"24 + 0 to 36 + 4 weeks' gestation were \"\n",
      "                                      'immunostained for intercellular '\n",
      "                                      'adhesion molecule-1, endothelial '\n",
      "                                      'leukocyte adhesion molecule-1, vascular '\n",
      "                                      'cell adhesion molecule-1, and platelet '\n",
      "                                      'endothelial cell adhesion molecule-1, '\n",
      "                                      'and the concentrations measured by '\n",
      "                                      'enzyme immunoassay. The results were '\n",
      "                                      'analyzed in relation to the degree of '\n",
      "                                      'cervical dilatation (< 2 cm, 2',\n",
      "                                'AD': 'Department of Obstetrics and '\n",
      "                                      'Gynecology, Freiburg University Medical '\n",
      "                                      'Center, Germany.',\n",
      "                                'CY': 'Germany',\n",
      "                                'DA': '20020104',\n",
      "                                'DCOM': '20020516',\n",
      "                                'DP': '2001',\n",
      "                                'EDAT': '2002/01/05 10:00',\n",
      "                                'FAU': ['Fischer, D C',\n",
      "                                        'Winkler, M',\n",
      "                                        'Ruck, P',\n",
      "                                        'Poth, D',\n",
      "                                        'Kemp, B',\n",
      "                                        'Rath, W'],\n",
      "                                'IP': '6',\n",
      "                                'IS': '0300-5577',\n",
      "                                'JID': '0361031',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['Adult',\n",
      "                                       'Antigens, CD31/analysis',\n",
      "                                       'Biopsy',\n",
      "                                       'Capillaries',\n",
      "                                       'Cell Adhesion Molecules/*analysis',\n",
      "                                       'Cesarean Section',\n",
      "                                       'E-Selectin/analysis',\n",
      "                                       'Endothelium, Vascular/chemistry',\n",
      "                                       'Female',\n",
      "                                       'Gestational Age',\n",
      "                                       'Human',\n",
      "                                       'Immunohistochemistry',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/analysis',\n",
      "                                       'Labor Stage, First',\n",
      "                                       'Labor, Premature/*metabolism',\n",
      "                                       'Linear Models',\n",
      "                                       'Pregnancy',\n",
      "                                       \"Support, Non-U.S. Gov't\",\n",
      "                                       'Up-Regulation',\n",
      "                                       'Uterus/blood supply/*chemistry',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/analysis'],\n",
      "                                'MHDA': '2002/05/17 10:01',\n",
      "                                'PG': '497-505',\n",
      "                                'PMID': '11776680',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (Antigens, CD31)',\n",
      "                                       '0 (Cell Adhesion Molecules)',\n",
      "                                       '0 (E-Selectin)',\n",
      "                                       '0 (Vascular Cell Adhesion Molecule-1)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'J Perinat Med 2001;29(6):497-505.',\n",
      "                                'TA': 'J Perinat Med',\n",
      "                                'TI': 'Localization and quantification of '\n",
      "                                      'adhesion molecule expression in the '\n",
      "                                      'lower uterine segment during premature '\n",
      "                                      'labor.',\n",
      "                                'UI': '21633239',\n",
      "                                'VI': '29'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '516927',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 27.396208,\n",
      "                    '_source': {'AB': 'OBJECTIVES: This study evaluated serum '\n",
      "                                      'concentrations of soluble cell adhesion '\n",
      "                                      'molecules in patients with gastric '\n",
      "                                      'cancer and in healthy control subjects. '\n",
      "                                      'Our objectives were to correlate these '\n",
      "                                      'levels with clinicopathological '\n",
      "                                      'features, established tumor markers, '\n",
      "                                      'and patient survival, and to assess '\n",
      "                                      'changes in serum levels of cell '\n",
      "                                      'adhesion molecules after tumor surgery. '\n",
      "                                      'METHODS: The serum concentrations of '\n",
      "                                      'the adhesion molecules E-selectin, '\n",
      "                                      'intercellular adhesion molecule-1 '\n",
      "                                      '(ICAM-1), and vascular cell adhesion '\n",
      "                                      'molecule-1 (VCAM-1) were investigated '\n",
      "                                      'by ELISA in 57 gastric cancer patients, '\n",
      "                                      'both before and 7 days after surgery, '\n",
      "                                      'and in 47 healthy control subjects. '\n",
      "                                      'RESULTS: Preoperative serum '\n",
      "                                      'concentrations of ICAM-1 and VCAM-1 in '\n",
      "                                      'gastric cancer patients were '\n",
      "                                      'significantly higher when compared with '\n",
      "                                      'those of healthy controls, whereas '\n",
      "                                      'there were no differences regarding '\n",
      "                                      'serum E-selectin levels. Serum levels '\n",
      "                                      'of E-selectin, ICAM-1, and VCAM-1 '\n",
      "                                      'correlated significantly with each '\n",
      "                                      'other. There was a significant '\n",
      "                                      'association between preoperative levels '\n",
      "                                      'of all three adhesion molecules and '\n",
      "                                      'disease stage, gastric wall invasion, '\n",
      "                                      'lymph node involvement, and presence of '\n",
      "                                      'distant metastases. Their '\n",
      "                                      'concentrations decreased significantly '\n",
      "                                      'after radical resection of the tumor, '\n",
      "                                      'whereas they remained almost unchanged '\n",
      "                                      'in patients with unresectable disease. '\n",
      "                                      'Elevated preoperative serum levels of '\n",
      "                                      'E-selectin, ICAM-1, and VCAM-1 levels '\n",
      "                                      'were found in 24.6%, 33.3%, and 28.1% '\n",
      "                                      'of patients, respectively. Elevated '\n",
      "                                      'levels of all three molecules were '\n",
      "                                      'significant prognostic factors for '\n",
      "                                      'patient survival but not independent of '\n",
      "                                      'disease stage. CONCLUSIONS: These '\n",
      "                                      'findings suggest that serum '\n",
      "                                      'concentrations of E-selectin, ICAM-1, '\n",
      "                                      'and VCAM-1 may reflect tumor '\n",
      "                                      'progression and metastasis, and may be '\n",
      "                                      'clinically useful.',\n",
      "                                'AD': 'First Department of Surgery, University '\n",
      "                                      'of Athens, Medical School, Greece.',\n",
      "                                'AID': 'S0002927002059270 [pii]',\n",
      "                                'CY': 'United States',\n",
      "                                'DA': '20030219',\n",
      "                                'DCOM': '20030401',\n",
      "                                'DP': '2003 Feb',\n",
      "                                'EDAT': '2003/02/20 04:00',\n",
      "                                'FAU': ['Alexiou, Dimitrios',\n",
      "                                        'Karayiannakis, Anastasios J',\n",
      "                                        'Syrigos, Konstantinos N',\n",
      "                                        'Zbar, Andrew',\n",
      "                                        'Sekara, Elefteria',\n",
      "                                        'Michail, Panagiotis',\n",
      "                                        'Rosenberg, Theophilos',\n",
      "                                        'Diamantis, Theodoros'],\n",
      "                                'IP': '2',\n",
      "                                'IS': '0002-9270',\n",
      "                                'JID': '0421030',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['Aged',\n",
      "                                       'Case-Control Studies',\n",
      "                                       'E-Selectin/*blood',\n",
      "                                       'Enzyme-Linked Immunosorbent Assay',\n",
      "                                       'Female',\n",
      "                                       'Human',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/*blood',\n",
      "                                       'Male',\n",
      "                                       'Neoplasm Staging',\n",
      "                                       'Predictive Value of Tests',\n",
      "                                       'Stomach '\n",
      "                                       'Neoplasms/*blood/mortality/surgery',\n",
      "                                       \"Support, Non-U.S. Gov't\",\n",
      "                                       'Tumor Markers, Biological/blood',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/*blood'],\n",
      "                                'MHDA': '2003/04/02 05:00',\n",
      "                                'PG': '478-85',\n",
      "                                'PMID': '12591072',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (E-Selectin)',\n",
      "                                       '0 (Tumor Markers, Biological)',\n",
      "                                       '0 (Vascular Cell Adhesion Molecule-1)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'Am J Gastroenterol 2003 '\n",
      "                                      'Feb;98(2):478-85.',\n",
      "                                'TA': 'Am J Gastroenterol',\n",
      "                                'TI': 'Clinical significance of serum levels '\n",
      "                                      'of E-selectin, intercellular adhesion '\n",
      "                                      'molecule-1, and vascular cell adhesion '\n",
      "                                      'molecule-1 in gastric cancer patients.',\n",
      "                                'UI': '22478633',\n",
      "                                'VI': '98'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '222286',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 27.267605,\n",
      "                    '_source': {'AB': 'The polysialylated, embryonic form of '\n",
      "                                      'the neuronal cell adhesion molecule '\n",
      "                                      '(PSA-NCAM) is known to participate in a '\n",
      "                                      'whole series of synaptic rearrangements '\n",
      "                                      'even in adult animals. The possible '\n",
      "                                      'role of this molecule in neuroplastic '\n",
      "                                      'changes of the adult rat somatosensory '\n",
      "                                      'cortex induced by unilateral '\n",
      "                                      'transection of the infraorbital branch '\n",
      "                                      'of the trigeminal nerve was studied '\n",
      "                                      'with PSA-NCAM immunostaining at light '\n",
      "                                      'microscopic level. Two',\n",
      "                                'AD': 'Laboratory of Molecular Neurobiology, '\n",
      "                                      'Institute of Biophysics, Biological '\n",
      "                                      'Research Center, Szeged, Hungary.',\n",
      "                                'CY': 'Hungary',\n",
      "                                'DA': '20020614',\n",
      "                                'DCOM': '20021204',\n",
      "                                'DP': '2002',\n",
      "                                'EDAT': '2002/06/18 10:00',\n",
      "                                'FAU': ['Hoyk, Zsofia',\n",
      "                                        'Varga, Cs',\n",
      "                                        'Parducz, A'],\n",
      "                                'IP': '1-2',\n",
      "                                'IS': '0236-5383',\n",
      "                                'JID': '8404358',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['Animal',\n",
      "                                       'Immunohistochemistry',\n",
      "                                       'Nerve Regeneration',\n",
      "                                       'Neural Cell Adhesion Molecule '\n",
      "                                       'L1/*metabolism',\n",
      "                                       'Neuronal Plasticity',\n",
      "                                       'Rats',\n",
      "                                       'Sialic Acids/*metabolism',\n",
      "                                       'Somatosensory Cortex/*metabolism',\n",
      "                                       \"Support, Non-U.S. Gov't\",\n",
      "                                       'Trigeminal '\n",
      "                                       'Nerve/*injuries/*metabolism'],\n",
      "                                'MHDA': '2002/12/05 04:00',\n",
      "                                'PG': '67-75',\n",
      "                                'PMID': '12064780',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (Neural Cell Adhesion Molecule L1)',\n",
      "                                       '0 (Sialic Acids)',\n",
      "                                       '0 (polysialyl neural cell adhesion '\n",
      "                                       'molecule)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'Acta Biol Hung 2002;53(1-2):67-75.',\n",
      "                                'TA': 'Acta Biol Hung',\n",
      "                                'TI': 'Transneuronal induction of the highly '\n",
      "                                      'sialylated isoform of the neural cell '\n",
      "                                      'adhesion molecule following nerve '\n",
      "                                      'injury.',\n",
      "                                'UI': '22059405',\n",
      "                                'VI': '53'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '472129',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 26.626034,\n",
      "                    '_source': {'AB': 'Intercellular adhesion molecule-1 '\n",
      "                                      '(ICAM-1) belongs to the superfamily of '\n",
      "                                      'immunoglobulin-like adhesion molecules. '\n",
      "                                      'Up-regulation of ICAM-1 occurs in many '\n",
      "                                      'different pathophysiological processes. '\n",
      "                                      'Also, cardiomyocytes can express '\n",
      "                                      'ICAM-1-for example, in acute myocardial '\n",
      "                                      'infarction. Moreover, inhibition of '\n",
      "                                      'ICAM-1 expression in the heart '\n",
      "                                      'dramatically reduces infarct size. '\n",
      "                                      'Hence, inhibitors of ICAM-1 may provide '\n",
      "                                      'a novel therapeutic option for acute '\n",
      "                                      'myocardial infarction.',\n",
      "                                'AD': 'Department of Pathology, ICaR-VU, VU '\n",
      "                                      'University Medical Center, Amsterdam, '\n",
      "                                      'the Netherlands. jmw.niessen@vumc.nl',\n",
      "                                'CY': 'United States',\n",
      "                                'DA': '20021217',\n",
      "                                'DCOM': '20030213',\n",
      "                                'DP': '2002 Nov',\n",
      "                                'EDAT': '2002/12/18 04:00',\n",
      "                                'FAU': ['Niessen, Hans W M',\n",
      "                                        'Krijnen, Paul A J',\n",
      "                                        'Visser, Cees A',\n",
      "                                        'Meijer, Chris J L M',\n",
      "                                        'Hack, C Erik'],\n",
      "                                'IS': '0077-8923',\n",
      "                                'JID': '7506858',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['Animal',\n",
      "                                       'Cell Adhesion',\n",
      "                                       'Cell Line',\n",
      "                                       'Gene Expression Regulation',\n",
      "                                       'Heart/physiology/*physiopathology',\n",
      "                                       'Human',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/*genetics',\n",
      "                                       'Myocardial Infarction/genetics',\n",
      "                                       'Myocardium/metabolism',\n",
      "                                       \"Support, Non-U.S. Gov't\",\n",
      "                                       'Transcription Factors/metabolism'],\n",
      "                                'MHDA': '2003/02/14 04:00',\n",
      "                                'PG': '573-85',\n",
      "                                'PMID': '12485931',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Review, Academic',\n",
      "                                'RF': '98',\n",
      "                                'RN': ['0 (Transcription Factors)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'Ann N Y Acad Sci 2002 Nov;973:573-85.',\n",
      "                                'TA': 'Ann N Y Acad Sci',\n",
      "                                'TI': 'Intercellular adhesion molecule-1 in '\n",
      "                                      'the heart.',\n",
      "                                'UI': '22373031',\n",
      "                                'VI': '973'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '335067',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 26.22314,\n",
      "                    '_source': {'AB': 'PURPOSE: The therapeutic potential of '\n",
      "                                      'TA, an anti-inflammatory '\n",
      "                                      'glucocorticoid, for the treatment of '\n",
      "                                      'exudative retinopathy has been examined '\n",
      "                                      'in several independent clinical '\n",
      "                                      'studies. The modulation of permeability '\n",
      "                                      'and adhesion molecule expression of an '\n",
      "                                      'epithelial cell line has been described '\n",
      "                                      'in vitro, with the use of cytokines and '\n",
      "                                      'triamcinolone acetonide (TA). In the '\n",
      "                                      'current study, the influence of '\n",
      "                                      'proinflammatory cytokines and TA on '\n",
      "                                      'permeability and adhesion molecule '\n",
      "                                      'expression in human choroidal '\n",
      "                                      'endothelial cells (CECs) was '\n",
      "                                      'investigated. METHODS: Human CEC '\n",
      "                                      'isolates treated with IFNgamma, '\n",
      "                                      'TNFalpha, and TA were evaluated by flow '\n",
      "                                      'cytometry and immunocytochemistry for '\n",
      "                                      'expression of intercellular adhesion '\n",
      "                                      'molecule (ICAM)-1, vascular cell '\n",
      "                                      'adhesion molecule (VCAM)-1, and major '\n",
      "                                      'histocompatibility complex (MHC)-I and '\n",
      "                                      '-II. The effects of IFNgamma, TNFalpha, '\n",
      "                                      'and TA on paracellular permeability of '\n",
      "                                      'CEC monolayers were assessed in '\n",
      "                                      'transendothelial cell resistance (TER) '\n",
      "                                      'assays. RESULTS: Both IFNgamma and '\n",
      "                                      'TNFalpha significantly upregulated '\n",
      "                                      'expression of ICAM1 and MHC-I on CECs. '\n",
      "                                      'Expression of VCAM1 was induced after '\n",
      "                                      'stimulation with both IFNgamma and '\n",
      "                                      'TNFalpha, whereas expression of MHC-II '\n",
      "                                      'was induced only by stimulation with '\n",
      "                                      'IFNgamma. Cytokine-induced expression '\n",
      "                                      'of ICAM1, MHC-I, and MHC-II antigen by '\n",
      "                                      'CECs was significantly downregulated by '\n",
      "                                      'TA. IFNgamma stimulation also increased '\n",
      "                                      'permeability of CEC monolayers, whereas '\n",
      "                                      'subsequent TA treatment decreased '\n",
      "                                      'permeability of CEC monolayers. '\n",
      "                                      'CONCLUSIONS: Human CEC isolates provide '\n",
      "                                      'a useful in vitro model to study '\n",
      "                                      'choroidal neovascular membrane '\n",
      "                                      'characteristics and their potential '\n",
      "                                      'response to pro',\n",
      "                                'AD': 'Save Sight Institute and Department of '\n",
      "                                      'Clinical Ophthalmology, University of '\n",
      "                                      'Sydney, New South Wales, Australia. '\n",
      "                                      'ppenfold@eye.usyd.edu.au',\n",
      "                                'CY': 'United States',\n",
      "                                'DA': '20020830',\n",
      "                                'DCOM': '20020912',\n",
      "                                'DP': '2002 Sep',\n",
      "                                'EDAT': '2002/08/31 10:00',\n",
      "                                'FAU': ['Penfold, Philip L',\n",
      "                                        'Wen, Li',\n",
      "                                        'Madigan, Michele C',\n",
      "                                        'King, Nicholas J C',\n",
      "                                        'Provis, Jan M'],\n",
      "                                'IP': '9',\n",
      "                                'IS': '0146-0404',\n",
      "                                'JID': '7703701',\n",
      "                                'LA': 'eng',\n",
      "                                'MH': ['Adult',\n",
      "                                       'Choroid/*blood supply',\n",
      "                                       'Cytokines/*pharmacology',\n",
      "                                       'Electric Conductivity',\n",
      "                                       'Endothelium, Vascular/cytology/drug '\n",
      "                                       'effects/*metabolism',\n",
      "                                       'Flow Cytometry',\n",
      "                                       'Fluorescent Antibody Technique, '\n",
      "                                       'Indirect',\n",
      "                                       'Glucocorticoids, '\n",
      "                                       'Synthetic/*pharmacology',\n",
      "                                       'Histocompatibility Antigens Class '\n",
      "                                       'I/metabolism',\n",
      "                                       'Histocompatibility Antigens Class '\n",
      "                                       'II/metabolism',\n",
      "                                       'Human',\n",
      "                                       'Intercellular Adhesion '\n",
      "                                       'Molecule-1/*metabolism',\n",
      "                                       'Middle Age',\n",
      "                                       'Permeability/drug effects',\n",
      "                                       \"Support, Non-U.S. Gov't\",\n",
      "                                       'Triamcinolone Acetonide/*pharmacology',\n",
      "                                       'Up-Regulation',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/*metabolism'],\n",
      "                                'MHDA': '2002/09/13 10:01',\n",
      "                                'PG': '3125-30',\n",
      "                                'PMID': '12202538',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (Cytokines)',\n",
      "                                       '0 (Glucocorticoids, Synthetic)',\n",
      "                                       '0 (Histocompatibility Antigens Class '\n",
      "                                       'I)',\n",
      "                                       '0 (Histocompatibility Antigens Class '\n",
      "                                       'II)',\n",
      "                                       '0 (Vascular Cell Adhesion Molecule-1)',\n",
      "                                       '126547-89-5 (Intercellular Adhesion '\n",
      "                                       'Molecule-1)',\n",
      "                                       '76-25-5 (Triamcinolone Acetonide)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'Invest Ophthalmol Vis Sci 2002 '\n",
      "                                      'Sep;43(9):3125-30.',\n",
      "                                'TA': 'Invest Ophthalmol Vis Sci',\n",
      "                                'TI': 'Modulation of permeability and adhesion '\n",
      "                                      'molecule expression by human choroidal '\n",
      "                                      'endothelial cells.',\n",
      "                                'UI': '22191250',\n",
      "                                'VI': '43'},\n",
      "                    '_type': '_doc'},\n",
      "                   {'_id': '463331',\n",
      "                    '_ignored': ['AB.keyword'],\n",
      "                    '_index': 'genomics-base',\n",
      "                    '_score': 25.96112,\n",
      "                    '_source': {'AB': 'Clinical significance of a soluble '\n",
      "                                      'vascular adhesion-1 molecule (pVCAM-1) '\n",
      "                                      \"in Sneddon's syndrome (SS) was studied \"\n",
      "                                      'in 48 SS patients by examination of '\n",
      "                                      'serum pVCAM-1 level using enzyme '\n",
      "                                      'immunoassay (R&D, USA). High serum '\n",
      "                                      'concentration of pVCAM-1 registered in '\n",
      "                                      '62.6% of SS patients was associated '\n",
      "                                      'with extracerebral thromboses, not with '\n",
      "                                      'severity of cerebral vessel damage. '\n",
      "                                      'Lethal outcomes for 2 years of '\n",
      "                                      'follow-up occurred more frequently in '\n",
      "                                      'patients with a high basal level of '\n",
      "                                      'pVCAM-1 than in patients with normal '\n",
      "                                      'level. Thus, development of the '\n",
      "                                      'pathological process in SS is '\n",
      "                                      'associated with elevated concentration '\n",
      "                                      'of serum pVCAM-1 and may serve an '\n",
      "                                      'additional laboratory indicator of '\n",
      "                                      'developing extrabrain thrombotic '\n",
      "                                      'disorders and marker of unfavourable '\n",
      "                                      'prognosis.',\n",
      "                                'CY': 'Russia',\n",
      "                                'DA': '20021210',\n",
      "                                'DCOM': '20030321',\n",
      "                                'DP': '2002',\n",
      "                                'EDAT': '2002/12/11 04:00',\n",
      "                                'FAU': ['Kalashnikova, L A',\n",
      "                                        'Aleksandrova, E N',\n",
      "                                        'Samsonov, M Iu',\n",
      "                                        'Dobrynina, L A',\n",
      "                                        'Novikov, A A',\n",
      "                                        'Amirov, M K',\n",
      "                                        'Nasonov, E L'],\n",
      "                                'IP': '10',\n",
      "                                'IS': '0023-2149',\n",
      "                                'JID': '2985204R',\n",
      "                                'LA': 'rus',\n",
      "                                'MH': ['English Abstract',\n",
      "                                       'Female',\n",
      "                                       'Human',\n",
      "                                       'Male',\n",
      "                                       'Middle Age',\n",
      "                                       'Prognosis',\n",
      "                                       'Sneddon '\n",
      "                                       'Syndrome/*blood/complications/mortality',\n",
      "                                       'Statistics, Nonparametric',\n",
      "                                       'Thrombosis/blood/etiology',\n",
      "                                       'Vascular Cell Adhesion '\n",
      "                                       'Molecule-1/*blood'],\n",
      "                                'MHDA': '2003/03/22 04:00',\n",
      "                                'PG': '35-8',\n",
      "                                'PMID': '12471836',\n",
      "                                'PST': 'ppublish',\n",
      "                                'PT': 'Journal Article',\n",
      "                                'RN': ['0 (Vascular Cell Adhesion Molecule-1)'],\n",
      "                                'SB': 'IM',\n",
      "                                'SO': 'Klin Med (Mosk) 2002;80(10):35-8.',\n",
      "                                'TA': 'Klin Med (Mosk)',\n",
      "                                'TI': '[Soluble vascular adhesion-1 molecule '\n",
      "                                      'in Sneddon syndrome: clinical and '\n",
      "                                      'prognostic value]',\n",
      "                                'TT': 'Rastvorimaia sosudistaia molekula '\n",
      "                                      'adgezii-1 pri sindrome Sneddona: '\n",
      "                                      'klinicheskoe i prognosticheskoe '\n",
      "                                      'znachenie.',\n",
      "                                'UI': '22360197',\n",
      "                                'VI': '80'},\n",
      "                    '_type': '_doc'}],\n",
      "          'max_score': 31.376125,\n",
      "          'total': {'relation': 'eq', 'value': 3404}},\n",
      " 'timed_out': False,\n",
      " 'took': 39}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_364/3076840873.py:15: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  response = es.search(index='genomics-base', body=body)\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "import elasticsearch\n",
    "\n",
    "# your code below\n",
    "#BEGIN ANSWER\n",
    "from pprint import pprint\n",
    "body = {\n",
    "    \"query\": { \n",
    "        \"simple_query_string\" : {\n",
    "          \"query\": \"molecule\"\n",
    "        }    \n",
    "    }\n",
    "}\n",
    "response = es.search(index='genomics-base', body=body)\n",
    "pprint(response)\n",
    "\n",
    "#END ANSWER\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Python client library returns Python objects, that use [dictionaries](https://docs.python.org/3.6/tutorial/datastructures.html#dictionaries) and [lists](https://docs.python.org/3.6/tutorial/introduction.html#lists).\n",
    "Use a [for loop](https://docs.python.org/3.6/tutorial/controlflow.html#for-statements) to inspect each hit, and print the retrieved document's titles one by one. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of results: 3404\n"
     ]
    }
   ],
   "source": [
    "#example\n",
    "print(\"Number of results: {}\".format(response['hits']['total']['value']))\n",
    "# your code below\n",
    "# for i in response['hits']['hits']:\n",
    "#     print(\"Output:\", i['_source'], \"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 02.C: _Search using the Elasticsearch DSL_\n",
    "\n",
    "You will notice that the native query format of Elasticsearch can be quite verbose.\n",
    "Elasticsearh provides the Python library `elasticsearch_dsl` to write more concise Elasticsearch queries. \n",
    "This is only to simplify the syntax: the library still issues Elasticsearch queries.\n",
    "\n",
    "For example, a simple `multi_match` query looks as follows:\n",
    "```python\n",
    "query = {\n",
    "   \"query\": {\n",
    "       \"multi_match\": {}\n",
    "   }\n",
    "}\n",
    "```\n",
    "\n",
    "The same query can be created with the DSL as follows:\n",
    "```python\n",
    "query = Q(\"multi_match\")\n",
    "```\n",
    "\n",
    "Especially for more complicated boolean queries, to use the native query format can become complicated.\n",
    "Read more about the DSL [here](https://elasticsearch-dsl.readthedocs.io/en/latest/search_dsl.html)\n",
    "\n",
    "__1. Search for the query `molecule` and check whether you get the same number of results as for exercise 02.A(2).__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10489\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jarron/.local/lib/python3.10/site-packages/elasticsearch/connection/base.py:200: ElasticsearchWarning: this request accesses system indices: [.apm-agent-configuration, .apm-custom-link, .kibana_7.17.6_001, .kibana_task_manager_7.17.6_001, .tasks], but in a future major version, direct access to system indices will be prevented by default\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# your code here\n",
    "# BEGIN ANSWER\n",
    "s1 = Search(using=es)\n",
    "q = Q(\"multi_match\", query=\"molecule\")\n",
    "s1 = s1.query(q)\n",
    "print(s1.count())\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Search for the documents that contain the words `cell` AND `blood`, using the DSL library. Check whether you get the same number of results as for exercise 02.A(4).__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21186\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# your code here\n",
    "# BEGIN ANSWER\n",
    "s1 = Search(using=es)\n",
    "q = Q(\"multi_match\", query=\"cell\") & Q(\"multi_match\", query=\"blood\")\n",
    "s1 = s1.query(q)\n",
    "print(s1.count())\n",
    "# response = s1.execute()\n",
    "# for results in response:\n",
    "#     print(results.to_dict(), \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "##  Exercise 02.D: Making your own TREC run\n",
    "\n",
    "We will adopt a scientific approach to building search engines. That is, we are not only going to build a search engine and see that it works, but we are also going to _measure_ how well it works, by measuring the search engine's quality. We will adopt the method from the [Text Retrieval Conference](http://trec.nist.gov) (TREC). TREC provides researchers with test collections, that consists of 3 parts:\n",
    "\n",
    "1. the document collection (in our case a part of the MEDLINE database)\n",
    "2. the topics (which are natural language descriptions of what the user is searching for: you can think of the as the _queries_)\n",
    "3. the relevance judgments (for each topic, what documents are relevant)\n",
    "\n",
    "\n",
    "\n",
    "__Exercise: Complete the code of the Python function `make_trec_run()` that reads the topics [FIR-s05-training-queries-simple.txt](data01/FIR-s05-training-queries-simple.txt), and for each topic does a search using Elasticsearch.__ The program should output a file in the [TREC submission format](https://trec-core.github.io/2017/#submission-guidelines). We already provided the first  lines for this exercise, which include:\n",
    "\n",
    "1. Open the file `'run_file_name'`' for writing and call it `run_file`.\n",
    "2. Open the file `'topics_file_name'` for reading, call it `test_queries`.\n",
    "3. For each line in `test_queries`:\n",
    "4. Remove the newline using `strip()`, then split the string on the tab character (`'\\t'`). The first part of the line is now `qid` (the query identifier) and the last part is `query` (a textual description of the query).\n",
    "5. complete the Python program such that the correct TREC run file is written to `'run_file_name'`.\n",
    "\n",
    "> **Note**: Make sure you output the `PMID` (pubmed identifier) of the document `hit['_source']['PMID']`. Do **not** use the elasticsearch identifier `_id` because they do not match the document identifiers in the relevance judgements. They were randomly generated by Elasticsearch during indexing.\n",
    "\n",
    "\n",
    "__Make sure to search in the fiels `TI` and `AB`, which correspond to the title and abstract, respectivelt, of the scientific papers of the MEDLINE collection.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "baseline.run is created\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "def make_trec_run(es, topics_file_name, run_file_name, index_name, run_name=\"test\"):\n",
    "    with open(run_file_name, 'w') as run_file:\n",
    "        with open(topics_file_name, 'r') as test_queries:\n",
    "            for line in test_queries:\n",
    "                (qid, query) = line.strip().split('\\t')\n",
    "                # BEGIN ANSWER\n",
    "                s = Search(using=es, index=index_name)\n",
    "                q = Q(\"multi_match\", query=query, fields=[\"TI\", \"AB\"])\n",
    "                s = s.query(q)\n",
    "                response = s.execute()\n",
    "                rank = 1\n",
    "                for results in response:\n",
    "                    score = response.hits[0].meta.score\n",
    "                    id = results['PMID']\n",
    "                    run_file.write(str(\"{} Q0 {} {} {} {}\\n\".format(qid, id, rank, score, run_name)))\n",
    "                    rank += 1\n",
    "    print(run_file_name + \" is created\")                \n",
    "                # END ANSWER\n",
    "                \n",
    "# connect to ES server             \n",
    "es = elasticsearch.Elasticsearch('localhost')\n",
    "# Write the results of the queries contained in the topic file `'data/training-queries-simple.txt'` \n",
    "# to the run file `'baseline.run'`, and name this test as `test01`\n",
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'baseline.run', \"genomics-base\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this prints out (it is a shell command) the content of the file baseline.run \n",
    "# !cat baseline.run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Tip: Write a line to `run_file` using `run_file.write(line)`. \n",
    "> The newline character is: `'\\n'`. Before writing a number to\n",
    "> the file, cast it to a string using `str()`.\n",
    ">\n",
    "> The TREC Submission guidelines allow you to submit up to 1000\n",
    "> documents per topic. Keep this in mind!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 03: Search models \n",
    "\n",
    "\n",
    "<span style=\"background:red; color: white;\">__You are advised to work on this part after Lecture 02__</span>\n",
    "\n",
    "\n",
    "### Background\n",
    "The way documents are indexed influences the performance of the IR systems. \n",
    "Elasticsearch [Mappings](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/mapping.html) define how a document, and its properties (fields) are stored and indexed, but also provides tools to implement and execute different document similarity measures (i.e. search models).  When using a different configuration of an ElasticSearch Mapping, the document collection needs to be re-indexed (or a new index need to be created - use the functions we provided above to do that).\n",
    "\n",
    "> See again: [Index Settings and Mappings](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/indices-create-index.html).\n",
    "> _Note: the default model (similarity) in ElasticSearch is BM25. Different models need to be specified (see example)._\n",
    "\n",
    "For instance, we can add a new field `\"title-abstract\"` that uses the  [similarity measure](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/similarity.html) _Boolean_, and let it serve as an index for the fields `\"TI\"` and `\"AB\"` (title and abstract):\n",
    "\n",
    "> Plase note that if you want to use the `boolean` similarity for the single fields, you need to specify it for each field. Otherwise, the default BM25 will be used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jarron/.local/lib/python3.10/site-packages/elasticsearch/connection/base.py:200: ElasticsearchWarning: Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.17/security-minimal-setup.html to enable security.\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n",
      "/tmp/ipykernel_799/1573036711.py:24: DeprecationWarning: The 'body' parameter is deprecated for the 'create' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  es.indices.create(index=index_name, body=body)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [3], line 27\u001b[0m\n\u001b[1;32m      1\u001b[0m boolean \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m      2\u001b[0m   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msettings\u001b[39m\u001b[38;5;124m\"\u001b[39m : {\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;66;03m# a single shard, so we do not suffer from approximate document frequencies\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     23\u001b[0m   }\n\u001b[1;32m     24\u001b[0m }\n\u001b[1;32m     26\u001b[0m es \u001b[38;5;241m=\u001b[39m elasticsearch\u001b[38;5;241m.\u001b[39mElasticsearch(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlocalhost\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 27\u001b[0m \u001b[43mindex_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdata01/FIR-s05-medline.json\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgenomics-bool\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mboolean\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn [1], line 29\u001b[0m, in \u001b[0;36mindex_documents\u001b[0;34m(es, collection_file_name, index_name, body)\u001b[0m\n\u001b[1;32m     27\u001b[0m create_index(es, index_name, body)\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# bulk index the documents from file_name\u001b[39;00m\n\u001b[0;32m---> 29\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43melasticsearch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhelpers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbulk\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     30\u001b[0m \u001b[43m    \u001b[49m\u001b[43mes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[43m    \u001b[49m\u001b[43mread_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcollection_file_name\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunk_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2000\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrequest_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m30\u001b[39;49m\n\u001b[1;32m     35\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/helpers/actions.py:410\u001b[0m, in \u001b[0;36mbulk\u001b[0;34m(client, actions, stats_only, ignore_status, *args, **kwargs)\u001b[0m\n\u001b[1;32m    408\u001b[0m \u001b[39m# make streaming_bulk yield successful results so we can count them\u001b[39;00m\n\u001b[1;32m    409\u001b[0m kwargs[\u001b[39m\"\u001b[39m\u001b[39myield_ok\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[0;32m--> 410\u001b[0m \u001b[39mfor\u001b[39;00m ok, item \u001b[39min\u001b[39;00m streaming_bulk(\n\u001b[1;32m    411\u001b[0m     client, actions, ignore_status\u001b[39m=\u001b[39mignore_status, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs\n\u001b[1;32m    412\u001b[0m ):\n\u001b[1;32m    413\u001b[0m     \u001b[39m# go through request-response pairs and detect failures\u001b[39;00m\n\u001b[1;32m    414\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m ok:\n\u001b[1;32m    415\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m stats_only:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/helpers/actions.py:329\u001b[0m, in \u001b[0;36mstreaming_bulk\u001b[0;34m(client, actions, chunk_size, max_chunk_bytes, raise_on_error, expand_action_callback, raise_on_exception, max_retries, initial_backoff, max_backoff, yield_ok, ignore_status, *args, **kwargs)\u001b[0m\n\u001b[1;32m    326\u001b[0m     time\u001b[39m.\u001b[39msleep(\u001b[39mmin\u001b[39m(max_backoff, initial_backoff \u001b[39m*\u001b[39m \u001b[39m2\u001b[39m \u001b[39m*\u001b[39m\u001b[39m*\u001b[39m (attempt \u001b[39m-\u001b[39m \u001b[39m1\u001b[39m)))\n\u001b[1;32m    328\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 329\u001b[0m     \u001b[39mfor\u001b[39;00m data, (ok, info) \u001b[39min\u001b[39;00m \u001b[39mzip\u001b[39m(\n\u001b[1;32m    330\u001b[0m         bulk_data,\n\u001b[1;32m    331\u001b[0m         _process_bulk_chunk(\n\u001b[1;32m    332\u001b[0m             client,\n\u001b[1;32m    333\u001b[0m             bulk_actions,\n\u001b[1;32m    334\u001b[0m             bulk_data,\n\u001b[1;32m    335\u001b[0m             raise_on_exception,\n\u001b[1;32m    336\u001b[0m             raise_on_error,\n\u001b[1;32m    337\u001b[0m             ignore_status,\n\u001b[1;32m    338\u001b[0m             \u001b[39m*\u001b[39margs,\n\u001b[1;32m    339\u001b[0m             \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs\n\u001b[1;32m    340\u001b[0m         ),\n\u001b[1;32m    341\u001b[0m     ):\n\u001b[1;32m    343\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m ok:\n\u001b[1;32m    344\u001b[0m             action, info \u001b[39m=\u001b[39m info\u001b[39m.\u001b[39mpopitem()\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/helpers/actions.py:240\u001b[0m, in \u001b[0;36m_process_bulk_chunk\u001b[0;34m(client, bulk_actions, bulk_data, raise_on_exception, raise_on_error, ignore_status, *args, **kwargs)\u001b[0m\n\u001b[1;32m    236\u001b[0m     ignore_status \u001b[39m=\u001b[39m (ignore_status,)\n\u001b[1;32m    238\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    239\u001b[0m     \u001b[39m# send the actual request\u001b[39;00m\n\u001b[0;32m--> 240\u001b[0m     resp \u001b[39m=\u001b[39m client\u001b[39m.\u001b[39;49mbulk(\u001b[39m*\u001b[39;49margs, body\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\\n\u001b[39;49;00m\u001b[39m\"\u001b[39;49m\u001b[39m.\u001b[39;49mjoin(bulk_actions) \u001b[39m+\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39m\\n\u001b[39;49;00m\u001b[39m\"\u001b[39;49m, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    241\u001b[0m \u001b[39mexcept\u001b[39;00m TransportError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    242\u001b[0m     gen \u001b[39m=\u001b[39m _process_bulk_chunk_error(\n\u001b[1;32m    243\u001b[0m         error\u001b[39m=\u001b[39me,\n\u001b[1;32m    244\u001b[0m         bulk_data\u001b[39m=\u001b[39mbulk_data,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    247\u001b[0m         raise_on_error\u001b[39m=\u001b[39mraise_on_error,\n\u001b[1;32m    248\u001b[0m     )\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/client/utils.py:347\u001b[0m, in \u001b[0;36mquery_params.<locals>._wrapper.<locals>._wrapped\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    345\u001b[0m     \u001b[39mif\u001b[39;00m p \u001b[39min\u001b[39;00m kwargs:\n\u001b[1;32m    346\u001b[0m         params[p] \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(p)\n\u001b[0;32m--> 347\u001b[0m \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, params\u001b[39m=\u001b[39;49mparams, headers\u001b[39m=\u001b[39;49mheaders, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/client/__init__.py:472\u001b[0m, in \u001b[0;36mElasticsearch.bulk\u001b[0;34m(self, body, index, doc_type, params, headers)\u001b[0m\n\u001b[1;32m    469\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mEmpty value passed for a required argument \u001b[39m\u001b[39m'\u001b[39m\u001b[39mbody\u001b[39m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    471\u001b[0m body \u001b[39m=\u001b[39m _bulk_body(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtransport\u001b[39m.\u001b[39mserializer, body)\n\u001b[0;32m--> 472\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mtransport\u001b[39m.\u001b[39;49mperform_request(\n\u001b[1;32m    473\u001b[0m     \u001b[39m\"\u001b[39;49m\u001b[39mPOST\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    474\u001b[0m     _make_path(index, doc_type, \u001b[39m\"\u001b[39;49m\u001b[39m_bulk\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[1;32m    475\u001b[0m     params\u001b[39m=\u001b[39;49mparams,\n\u001b[1;32m    476\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    477\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    478\u001b[0m )\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/transport.py:427\u001b[0m, in \u001b[0;36mTransport.perform_request\u001b[0;34m(self, method, url, headers, params, body)\u001b[0m\n\u001b[1;32m    424\u001b[0m connection \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_connection()\n\u001b[1;32m    426\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 427\u001b[0m     status, headers_response, data \u001b[39m=\u001b[39m connection\u001b[39m.\u001b[39;49mperform_request(\n\u001b[1;32m    428\u001b[0m         method,\n\u001b[1;32m    429\u001b[0m         url,\n\u001b[1;32m    430\u001b[0m         params,\n\u001b[1;32m    431\u001b[0m         body,\n\u001b[1;32m    432\u001b[0m         headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    433\u001b[0m         ignore\u001b[39m=\u001b[39;49mignore,\n\u001b[1;32m    434\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[1;32m    435\u001b[0m     )\n\u001b[1;32m    437\u001b[0m     \u001b[39m# Lowercase all the header names for consistency in accessing them.\u001b[39;00m\n\u001b[1;32m    438\u001b[0m     headers_response \u001b[39m=\u001b[39m {\n\u001b[1;32m    439\u001b[0m         header\u001b[39m.\u001b[39mlower(): value \u001b[39mfor\u001b[39;00m header, value \u001b[39min\u001b[39;00m headers_response\u001b[39m.\u001b[39mitems()\n\u001b[1;32m    440\u001b[0m     }\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/connection/http_urllib3.py:255\u001b[0m, in \u001b[0;36mUrllib3HttpConnection.perform_request\u001b[0;34m(self, method, url, params, body, timeout, ignore, headers)\u001b[0m\n\u001b[1;32m    252\u001b[0m     body \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_gzip_compress(body)\n\u001b[1;32m    253\u001b[0m     request_headers[\u001b[39m\"\u001b[39m\u001b[39mcontent-encoding\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mgzip\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 255\u001b[0m response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpool\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    256\u001b[0m     method, url, body, retries\u001b[39m=\u001b[39;49mRetry(\u001b[39mFalse\u001b[39;49;00m), headers\u001b[39m=\u001b[39;49mrequest_headers, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkw\n\u001b[1;32m    257\u001b[0m )\n\u001b[1;32m    258\u001b[0m response_headers \u001b[39m=\u001b[39m {\n\u001b[1;32m    259\u001b[0m     header\u001b[39m.\u001b[39mlower(): value \u001b[39mfor\u001b[39;00m header, value \u001b[39min\u001b[39;00m response\u001b[39m.\u001b[39mheaders\u001b[39m.\u001b[39mitems()\n\u001b[1;32m    260\u001b[0m }\n\u001b[1;32m    261\u001b[0m duration \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mtime() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:665\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    662\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_prepare_proxy(conn)\n\u001b[1;32m    664\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 665\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    666\u001b[0m     conn,\n\u001b[1;32m    667\u001b[0m     method,\n\u001b[1;32m    668\u001b[0m     url,\n\u001b[1;32m    669\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    670\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    671\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    672\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    673\u001b[0m )\n\u001b[1;32m    675\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    676\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    677\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    678\u001b[0m \u001b[39m# mess.\u001b[39;00m\n\u001b[1;32m    679\u001b[0m response_conn \u001b[39m=\u001b[39m conn \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m release_conn \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:421\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    416\u001b[0m             httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39mgetresponse()\n\u001b[1;32m    417\u001b[0m         \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    418\u001b[0m             \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    419\u001b[0m             \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    420\u001b[0m             \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[0;32m--> 421\u001b[0m             six\u001b[39m.\u001b[39;49mraise_from(e, \u001b[39mNone\u001b[39;49;00m)\n\u001b[1;32m    422\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    423\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_raise_timeout(err\u001b[39m=\u001b[39me, url\u001b[39m=\u001b[39murl, timeout_value\u001b[39m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:416\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    413\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[1;32m    414\u001b[0m     \u001b[39m# Python 3\u001b[39;00m\n\u001b[1;32m    415\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 416\u001b[0m         httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    417\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    418\u001b[0m         \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    419\u001b[0m         \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    420\u001b[0m         \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[1;32m    421\u001b[0m         six\u001b[39m.\u001b[39mraise_from(e, \u001b[39mNone\u001b[39;00m)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1374\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1372\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m   1373\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1374\u001b[0m         response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1375\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n\u001b[1;32m   1376\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mclose()\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[39m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    319\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n\u001b[1;32m    320\u001b[0m         \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_read_status\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfp\u001b[39m.\u001b[39;49mreadline(_MAXLINE \u001b[39m+\u001b[39;49m \u001b[39m1\u001b[39;49m), \u001b[39m\"\u001b[39m\u001b[39miso-8859-1\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(line) \u001b[39m>\u001b[39m _MAXLINE:\n\u001b[1;32m    281\u001b[0m         \u001b[39mraise\u001b[39;00m LineTooLong(\u001b[39m\"\u001b[39m\u001b[39mstatus line\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m    704\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sock\u001b[39m.\u001b[39;49mrecv_into(b)\n\u001b[1;32m    706\u001b[0m     \u001b[39mexcept\u001b[39;00m timeout:\n\u001b[1;32m    707\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_timeout_occurred \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "boolean = {\n",
    "  \"settings\" : {\n",
    "    # a single shard, so we do not suffer from approximate document frequencies\n",
    "    \"number_of_shards\" : 1\n",
    "  },\n",
    "  \"mappings\": {\n",
    "      \"properties\": {\n",
    "        \"AB\": {\n",
    "          \"type\": \"text\",\n",
    "          \"copy_to\": \"title-abstract\",\n",
    "          \"similarity\": \"boolean\"\n",
    "        },\n",
    "        \"TI\": {\n",
    "          \"type\": \"text\",\n",
    "          \"copy_to\": \"title-abstract\",\n",
    "          \"similarity\": \"boolean\"\n",
    "        },\n",
    "        \"title-abstract\": {  # compound field\n",
    "          \"type\": \"text\",\n",
    "          \"similarity\": \"boolean\"\n",
    "        }\n",
    "      }\n",
    "  }\n",
    "}\n",
    "\n",
    "es = elasticsearch.Elasticsearch('localhost')\n",
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-bool', body=boolean)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Most changes to the mappings cannot be done on an existing index. Some (for instance\n",
    "> similarity measures) can be changed if the index is first closed. Nevertheless, we \n",
    "> will in this notebook _re-index_ the collection for every change to the mappings\n",
    "> using the function `index_documents()` that we defined above. Mappings (and settings)\n",
    "> can be passed to the function using the `body` parameter.\n",
    "\n",
    "<span style=\"background:#444; color: white;\">__We suggest you to create different indices using different models of search (according to the available disk space on your VM). This will avoid that changes are not correctly applied, and you won't see the expected results.__</span>\n",
    "\n",
    "<span style=\"background:#444; color: white;\">E.g. for the 'boolean' model, we created the 'genomics-bool' index.</span>\n",
    "\n",
    "Let's have a look at the mappings and settings for our index as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genomics-bool': {'aliases': {},\n",
       "  'mappings': {'properties': {'AB': {'type': 'text',\n",
       "     'copy_to': ['title-abstract'],\n",
       "     'similarity': 'boolean'},\n",
       "    'AD': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'AID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CON': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'CY': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'DA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'DCOM': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'DP': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'EDAT': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'EIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'FAU': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'ID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'IP': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'IS': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'JID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'LA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'LR': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'MH': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'MHDA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PG': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PHST': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PMID': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PS': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PST': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'PT': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RF': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'RPI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SB': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SO': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'SPIN': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'TA': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'TI': {'type': 'text',\n",
       "     'copy_to': ['title-abstract'],\n",
       "     'similarity': 'boolean'},\n",
       "    'TT': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'UI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'VI': {'type': 'text',\n",
       "     'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}},\n",
       "    'title-abstract': {'type': 'text', 'similarity': 'boolean'}}},\n",
       "  'settings': {'index': {'routing': {'allocation': {'include': {'_tier_preference': 'data_content'}}},\n",
       "    'number_of_shards': '1',\n",
       "    'provided_name': 'genomics-bool',\n",
       "    'creation_date': '1664316713041',\n",
       "    'number_of_replicas': '1',\n",
       "    'uuid': '8Z8VO0VRScu4-dmtGXNH1A',\n",
       "    'version': {'created': '7170699'}}}}}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "es.indices.get(index='genomics-bool')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's search our new field `\"title-abstract\"` as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_364/3049803502.py:9: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  es.search(index=\"genomics-bool\", search_type=search_type, body=body)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'took': 18,\n",
       " 'timed_out': False,\n",
       " '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0},\n",
       " 'hits': {'total': {'value': 263, 'relation': 'eq'},\n",
       "  'max_score': 1.0,\n",
       "  'hits': [{'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '125',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Previous studies from our laboratories revealed the susceptibility of Leishmania sp. to glibenclamide (GLIB), a potassium channel blocker which selectively interacts with adenosine-binding-cassette transporters. In the present work, we analyzed whether the drug sensitivity of intracellular amastigotes correlates with changes in macrophage features that are related to their function as antigen-presenting cells. We provide evidence that in BALB/c murine macrophages, GLIB induced a decrease in the interferon-gamma-stimulated expression of major histocompatibility complex class II molecules and the co-stimulatory molecule CD86 (B7-2). Furthermore, it caused a decrease in the interleukin-1 secretion by macrophages. The data indicate that the treatment with GLIB inhibits the Th2 development and polarizes macrophage functions towards the induction of a protective Th1 response.',\n",
       "     'AD': 'Laboratory of Molecular Physiology, Instituto de Medicina Experimental, Facultad de Medicina, Universidad Central de Venezuela, Caracas. aiponte@reacciun.ve',\n",
       "     'CY': 'United States',\n",
       "     'DA': '20011105',\n",
       "     'DCOM': '20020412',\n",
       "     'DP': '2001 Aug',\n",
       "     'EDAT': '2001/11/06 10:00',\n",
       "     'IP': '3',\n",
       "     'IS': '0892-3973',\n",
       "     'JID': '8800150',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/16 10:01',\n",
       "     'PG': '477-86',\n",
       "     'PMID': '11694036',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Immunopharmacol Immunotoxicol 2001 Aug;23(3):477-86.',\n",
       "     'TA': 'Immunopharmacol Immunotoxicol',\n",
       "     'TI': 'Experimental leishmaniasis: the glibenclamide-triggered decrease in parasite growth correlates with changes in macrophage features.',\n",
       "     'UI': '21545398',\n",
       "     'VI': '23',\n",
       "     'MH': ['Animal',\n",
       "      'Antigens, CD/metabolism',\n",
       "      'Glyburide/*therapeutic use',\n",
       "      'Histocompatibility Antigens Class II/metabolism',\n",
       "      'In Vitro',\n",
       "      'Interleukin-1/metabolism',\n",
       "      'Leishmania major/drug effects/growth & development',\n",
       "      'Leishmaniasis/*drug therapy/*immunology/parasitology',\n",
       "      'Macrophages/*drug effects/*immunology/parasitology',\n",
       "      'Membrane Glycoproteins/metabolism',\n",
       "      'Mice',\n",
       "      'Mice, Inbred BALB C',\n",
       "      \"Support, Non-U.S. Gov't\"],\n",
       "     'RN': ['0 (Antigens, CD)',\n",
       "      '0 (B7-2 protein)',\n",
       "      '0 (Histocompatibility Antigens Class II)',\n",
       "      '0 (Interleukin-1)',\n",
       "      '0 (Membrane Glycoproteins)',\n",
       "      '10238-21-8 (Glyburide)'],\n",
       "     'FAU': ['Ponte-Sucre, A', 'Figarella, K', 'Moll, H']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '639',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Osteoblast differentiation is a multistep process that involves critical spatial and temporal regulation of cellular processes marked by the presence of a large number of differentially expressed molecules. To identify key functional molecules, we used differential messenger RNA (mRNA) display and compared RNA populations isolated from the defined transition phases (proliferation, matrix formation, and mineralization) of the MC3T3-E1 osteoblast-like cell line. Using this approach, a complementary DNA (cDNA) fragment was isolated and identified as neuroleukin (NLK), a multifunctional cytokine also known as autocrine motility factor (AMF), phosphoglucose isomerase (PGI; phosphohexose isomerase [PHI]), and maturation factor (MF). Northern analysis showed NLK temporal expression during MC3T3-E1 cell differentiation with a 3.5-fold increase during matrix formation and mineralization. Immunocytochemical studies revealed the presence of NLK in MC3T3-E1 cells as well as in the surrounding matrix, consistent with a secreted molecule. In contrast, the NLK receptor protein was detected primarily on the cell membrane. In subsequent studies, a high level of NLK expression was identified in osteoblasts and superficial articular chondrocytes in bone of 1-, 4-, and 8-month-old normal mice, as well as in fibroblasts, proliferating chondrocytes, and osteoblasts within a fracture callus. However, NLK was not evident in hypertrophic chondrocytes or osteocytes. In addition, treatment of MC3T3 cells with 6-phosphogluconic acid (6PGA; a NLK inhibitor) resulted in diminishing alkaline phosphatase (ALP) activity and mineralization in MC3T3-E1 cells, especially during the matrix formation stage of differentiating cells. Taken together, these data show specific expression of NLK in discrete populations of bone and cartilage cells and suggest a possible role for this secreted protein in bone development and regeneration.',\n",
       "     'AD': 'Department of Biomedical Engineering, State University of New York at Stony Brook, 11794-2580, USA.',\n",
       "     'CY': 'United States',\n",
       "     'DA': '20011107',\n",
       "     'DCOM': '20020422',\n",
       "     'DP': '2001 Nov',\n",
       "     'EDAT': '2001/11/08 10:00',\n",
       "     'IP': '11',\n",
       "     'IS': '0884-0431',\n",
       "     'JID': '8610640',\n",
       "     'LA': 'eng',\n",
       "     'LR': '20020610',\n",
       "     'MHDA': '2002/04/23 10:01',\n",
       "     'PG': '1994-2004',\n",
       "     'PMID': '11697795',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'S',\n",
       "     'SO': 'J Bone Miner Res 2001 Nov;16(11):1994-2004.',\n",
       "     'TA': 'J Bone Miner Res',\n",
       "     'TI': 'Differential expression of neuroleukin in osseous tissues and its involvement in mineralization during osteoblast differentiation.',\n",
       "     'UI': '21554149',\n",
       "     'VI': '16',\n",
       "     'MH': ['3T3 Cells',\n",
       "      'Animal',\n",
       "      'Bone Development/genetics/physiology',\n",
       "      'Cell Differentiation/drug effects',\n",
       "      'Fracture Healing/genetics/physiology',\n",
       "      'Gene Expression',\n",
       "      'Gluconates/pharmacology',\n",
       "      'Immunohistochemistry',\n",
       "      'Mice',\n",
       "      'Minerals/metabolism',\n",
       "      'Neuroleukin/antagonists & inhibitors/*genetics',\n",
       "      'Osteoblasts/*cytology/drug effects/*metabolism',\n",
       "      'RNA, Messenger/genetics/metabolism',\n",
       "      'Receptors, Growth Factor/metabolism',\n",
       "      \"Support, Non-U.S. Gov't\",\n",
       "      \"Support, U.S. Gov't, Non-P.H.S.\"],\n",
       "     'RN': ['0 (Gluconates)',\n",
       "      '0 (Minerals)',\n",
       "      '0 (Neuroleukin)',\n",
       "      '0 (RNA, Messenger)',\n",
       "      '0 (Receptors, Growth Factor)',\n",
       "      '921-62-0 (6-phosphogluconate)'],\n",
       "     'FAU': ['Zhi, J', 'Sommerfeldt, D W', 'Rubin, C T', 'Hadjiargyrou, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '713',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'CD38 identifies a surface molecule with multi-functional activity. Its prognostic importance in B-cell chronic lymphocytic leukemia (B-CLL) is currently under investigation in view of the fact that two different groups have recently indicated that CD38 expression could be an independent prognostic marker in B-CLL. We analyzed the clinico-biological features of 61 immunologically typical (CD5+CD23+) B-CLL patients stratified according to the CD38 expression. Twenty-two (36%) patients expressed CD38 in more than 30% of CD19-positive cells and were considered as CD38-positive B-CLL. Atypical morphology (p 0.02), peripheral blood lymphocytosis (p 0.01) and diffuse histopathologic bone marrow pattern (p 0.003) were findings found to be closely associated with CD38 expression. On the other hand, A and B Binet stages (p 0.02) and interstitial bone marrow involvement (p 0.005) were more represented in the CD38-negative B-CLL group. Trisomy 12 was detected more frequently in the CD38-positive B-CLL group, while 13q14 deletions mainly occurred in CD38-negative group (p 0.005). Finally, median survival of CD38-positive B-CLL patients was 90 months, while it was not reached at 180 months in CD38-negative patients. Taken together, our data strongly suggest that the evaluation of CD38 expression may identify two groups patients with B-CLL greatly differing in their clinico-biological features.',\n",
       "     'AD': 'Division of Hematology, IRCCS Casa Sollievo della Sofferenza Hospital, 71013 San Giovanni Rotondo, Italy. ematologia@operapadrepio.it',\n",
       "     'CY': 'Switzerland',\n",
       "     'DA': '20011108',\n",
       "     'DCOM': '20020516',\n",
       "     'DP': '2001 Jun',\n",
       "     'EDAT': '2001/11/09 10:00',\n",
       "     'IP': '1-2',\n",
       "     'IS': '1042-8194',\n",
       "     'JID': '9007422',\n",
       "     'LA': 'eng',\n",
       "     'LR': '20021213',\n",
       "     'MHDA': '2002/05/17 10:01',\n",
       "     'PG': '109-14',\n",
       "     'PMID': '11699197',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Leuk Lymphoma 2001 Jun;42(1-2):109-14.',\n",
       "     'TA': 'Leuk Lymphoma',\n",
       "     'TI': 'CD38 expression correlates with adverse biological features and predicts poor clinical outcome in B-cell chronic lymphocytic leukemia.',\n",
       "     'UI': '21556526',\n",
       "     'VI': '42',\n",
       "     'MH': ['ADP-ribosyl Cyclase',\n",
       "      'Adult',\n",
       "      'Aged',\n",
       "      'Antigens, Differentiation/*biosynthesis',\n",
       "      'Antigens, Neoplasm/*biosynthesis',\n",
       "      'Biological Markers/analysis',\n",
       "      'Cohort Studies',\n",
       "      'Cytogenetic Analysis',\n",
       "      'Human',\n",
       "      'Leukemia, B-Cell, Chronic/diagnosis/*immunology/mortality',\n",
       "      'Middle Age',\n",
       "      'NAD+ Nucleosidase/*biosynthesis',\n",
       "      'Prognosis',\n",
       "      'Survival Analysis',\n",
       "      'Treatment Outcome'],\n",
       "     'RN': ['0 (Antigens, Differentiation)',\n",
       "      '0 (Antigens, Neoplasm)',\n",
       "      '0 (Biological Markers)',\n",
       "      'EC 3.2.2.5 (ADP-ribosyl Cyclase)',\n",
       "      'EC 3.2.2.5 (CD38 antigen)',\n",
       "      'EC 3.2.2.5 (NAD+ Nucleosidase)'],\n",
       "     'FAU': [\"D'Arena, G\",\n",
       "      'Musto, P',\n",
       "      'Cascavilla, N',\n",
       "      \"Dell'Olio, M\",\n",
       "      'Di Renzo, N',\n",
       "      'Perla, G',\n",
       "      'Savino, L',\n",
       "      'Carotenuto, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '737',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'The emigration of lymphocytes from blood into lymph nodes is regulated by the expression of the adhesion molecule, L-selectin on the lymphocyte surface which arrests the rolling of the cell on the vessel wall and allows firmer adhesive interactions to develop. The expression of L-selectin on B-CLL lymphocytes is less than half that on normal lymphocytes and this difference correlates with an impaired capacity of B-CLL lymphocytes to migrate beneath a monolayer of human umbilical vein endothelial cells. Both the B-cell and T-cell lymphocytes from normal subjects and B-CLL patients show down-regulation of L-selectin and CD23 after transendothelial migration. The reduced expression of L-selectin on B-CLL lymphocytes leads to a relative \"trapping\" of these cells in the vascular space and is one factor contributing to the elevation of peripheral lymphocyte count.',\n",
       "     'AD': 'Sydney University, Department of Medicine, Nepean Hospital, Penrith, NSW 2750, Australia.',\n",
       "     'CY': 'Switzerland',\n",
       "     'DA': '20011108',\n",
       "     'DCOM': '20020516',\n",
       "     'DP': '2001 Jun',\n",
       "     'EDAT': '2001/11/09 10:00',\n",
       "     'IP': '1-2',\n",
       "     'IS': '1042-8194',\n",
       "     'JID': '9007422',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/05/17 10:01',\n",
       "     'PG': '5-12',\n",
       "     'PMID': '11699221',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Review, Tutorial',\n",
       "     'RF': '34',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Leuk Lymphoma 2001 Jun;42(1-2):5-12.',\n",
       "     'TA': 'Leuk Lymphoma',\n",
       "     'TI': 'Impaired transendothelial migration of B-CLL lymphocytes: a defect linked to low L-selectin expression.',\n",
       "     'UI': '21556515',\n",
       "     'VI': '42',\n",
       "     'MH': ['B-Lymphocytes/pathology',\n",
       "      'Chemotaxis, Leukocyte',\n",
       "      'Down-Regulation',\n",
       "      'Endothelium, Vascular/*physiopathology',\n",
       "      'Human',\n",
       "      'L-Selectin/biosynthesis/physiology',\n",
       "      'Leukemia, B-Cell, Chronic/*pathology/physiopathology'],\n",
       "     'RN': ['126880-86-2 (L-Selectin)'],\n",
       "     'FAU': ['Gu, B', 'Dao, L P', 'Wiley, J']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '858',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Monolayers of seven fractions of natural lipids (phosphatidyl inositol, sulfoquinovosyl diacylglycerol, phosphatidylcholine, digalactosyl diacylglycerol, phosphatidyl glycerol, phosphatidylethanolamine, monogalactosyl diacylglycerol), isolated from the photoautotrophic cell culture of the moss Marchantia polymorpha grown under normal and light-stress conditions, have been prepared for the first time. We have shown that the high-intensity light affects the area occupied by the lipid molecule. In the case of digalactosyl diacylglycerol and phosphatidyl glycerol fractions, after the light stress the area significantly increased from 0.50 to 0.80 nm2 and from 0.47 to 0.63 nm2, respectively, and in the case of the sulfoquinovosyl diacylglycerol fraction, the area decreased from 0.40 to 0.32 nm2. These results are in agreement with our previous data on the redistribution of the double bonds in the aliphatic chains of these lipids and can be used to characterize the state of the lipid bilayer of the thylakoid membranes.',\n",
       "     'AD': 'Shemyakin and Ovchinnikov Institute of Bioorganic Chemistry, Russian Academy of Sciences, Moscow.',\n",
       "     'CY': 'Switzerland',\n",
       "     'DA': '20011108',\n",
       "     'DCOM': '20020411',\n",
       "     'DP': '2001 Jul',\n",
       "     'EDAT': '2001/11/09 10:00',\n",
       "     'IP': '5',\n",
       "     'IS': '1023-6597',\n",
       "     'JID': '9517472',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/12 10:01',\n",
       "     'PG': '579-85',\n",
       "     'PMID': '11699862',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Membr Cell Biol 2001 Jul;14(5):579-85.',\n",
       "     'TA': 'Membr Cell Biol',\n",
       "     'TI': 'A study of moss (Marchantia polymorpha) thylakoid membrane lipids in monolayers.',\n",
       "     'UI': '21556221',\n",
       "     'VI': '14',\n",
       "     'MH': ['Cells, Cultured/chemistry/radiation effects',\n",
       "      'Chromatography, Thin Layer',\n",
       "      'Diglycerides/chemistry/isolation & purification',\n",
       "      'Glycolipids/chemistry/isolation & purification',\n",
       "      'Light',\n",
       "      'Membrane Lipids/*chemistry/isolation & purification',\n",
       "      'Mosses/chemistry/radiation effects',\n",
       "      'Phosphatidylcholines/chemistry/isolation & purification',\n",
       "      'Phosphatidylethanolamines/chemistry/isolation & purification',\n",
       "      'Phosphatidylglycerols/chemistry/isolation & purification',\n",
       "      'Phosphatidylinositols/chemistry/isolation & purification',\n",
       "      'Plants/*chemistry/radiation effects',\n",
       "      \"Support, Non-U.S. Gov't\",\n",
       "      'Thylakoids/*chemistry/radiation effects'],\n",
       "     'RN': ['0 (Diglycerides)',\n",
       "      '0 (Glycolipids)',\n",
       "      '0 (Membrane Lipids)',\n",
       "      '0 (Phosphatidylcholines)',\n",
       "      '0 (Phosphatidylethanolamines)',\n",
       "      '0 (Phosphatidylglycerols)',\n",
       "      '0 (Phosphatidylinositols)',\n",
       "      '0 (digalactosyldiglycerides)',\n",
       "      '0 (monogalactosyldiglycerides)',\n",
       "      '0 (sulfoquinovosyl diglyceride)'],\n",
       "     'FAU': ['Volchenkova, T A',\n",
       "      'Kalabina, N A',\n",
       "      'Schaefer, C',\n",
       "      'Zubov, V P']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '1206',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Photodynamic therapy (PDT) is the treatment of tumors or dysplasic tissue with drugs that produce cytotoxic metabolites when exposed to light. Aminolevulinic acid HCl (5-aminolevulinic acid HCl; ALA) is a prodrug that is metabolized intracellularly to form the photosensitizing molecule protoporphyrin (PpIX). When PpIX is activated by light, cytotoxic reactive oxygen species and free radicals are generated. ALA can diffuse through skin and preferentially localizes in tumors and dysplasic tissue; subsequent exposure of PpIX-loaded tumor cells to light can destroy the tumor. After application of a 20% solution of ALA to actinic keratosis lesions of the head, PpIX (as measured by skin fluorescence) peaked 11 hours after treatment and the mean clearance half-life was 30 hours. In phase II trials 10 J/cm2 of blue light (wavelength = 417 nm) delivered at 10 mW/cm2 for 1000 seconds was found to provide maximal therapeutic effect on lesions of the head after treatment with 20% ALA. In phase III trials of ALA PDT in 241 patients with lesions of the head 72% of patients had a complete response to treatment at 12 weeks versus 20% of those treated with vehicle and light alone. Some of these patients had been re-treated at 8 weeks. In these trials 12% of ALA-treated patients and 37.5% of those receiving vehicle whose lesions had cleared at 8 weeks had relapsed at 12 weeks. When the total number of lesions were considered the recurrence rate was 5 and 27.9% for ALA',\n",
       "     'AD': 'Adis International Limited, Auckland, New Zealand. demail@adis.co.nz',\n",
       "     'CY': 'New Zealand',\n",
       "     'DA': '20011112',\n",
       "     'DCOM': '20020716',\n",
       "     'DP': '2000 Mar-Apr',\n",
       "     'EDAT': '2001/11/13 10:00',\n",
       "     'IP': '2',\n",
       "     'IS': '1175-0561',\n",
       "     'JID': '100895290',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/07/18 10:01',\n",
       "     'PG': '133-9; discussion 140-1',\n",
       "     'PMID': '11702313',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Review, Tutorial',\n",
       "     'RF': '42',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Am J Clin Dermatol 2000 Mar-Apr;1(2):133-9; discussion 140-1.',\n",
       "     'TA': 'Am J Clin Dermatol',\n",
       "     'TI': 'Topical aminolevulinic acid HCl photodynamic therapy.',\n",
       "     'UI': '21559338',\n",
       "     'VI': '1',\n",
       "     'MH': ['Administration, Topical',\n",
       "      'Aminolevulinic Acid/*administration & dosage/adverse effects',\n",
       "      'Clinical Trials, Phase II',\n",
       "      'Clinical Trials, Phase III',\n",
       "      'Human',\n",
       "      '*Photochemotherapy',\n",
       "      'Photosensitizing Agents/*administration & dosage/adverse effects'],\n",
       "     'RN': ['0 (Photosensitizing Agents)', '106-60-5 (Aminolevulinic Acid)'],\n",
       "     'FAU': ['Ormrod, D', 'Jarvis, B']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '1684',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'We describe an approach for the rapid mapping of epitopes within a malaria antigen using a combination of phage display techniques. Phage display of antigen fragments identifies the location of the epitopes, then random peptide libraries displayed on phage are employed to identify accurately amino acids involved in the epitope. Finally, phage display of mutant fragments confirms the role of each residue in the epitope. This approach was applied to the apical membrane antigen-1 (AMA1), which is a leading candidate for inclusion in a vaccine directed against the asexual blood stages of Plasmodium falciparum. As part of the effort both to understand the function of AMA1 in the parasite life cycle and to define the specificity of protective immune responses, a panel of monoclonal antibodies (MAbs) was generated to obtain binding reagents to the various domains within the molecule. There is a pressing need to determine rapidly the regions recognized by these antibodies and the structural requirements required within AMA1 for high affinity binding of the MAbs. Using phage displaying random AMA1 fragments, it was shown that MAb5G8 recognizes a short linear epitope within the pro-domain of AMA1 whereas the epitope recognized by MAb 1F9 is reduction sensitive and resides within a disulphide-bonded 57 amino acid sub-domain of domain-1. Phage displaying random peptide libraries and mutant AMA1 fragments were employed for fine mapping of the MAb5G8 core epitope to a three-residue sequence in the AMA1 prodomain.',\n",
       "     'AD': 'Department of Biochemistry, La Trobe University, Bundoora, 3083, Victoria, Australia.',\n",
       "     'CY': 'England',\n",
       "     'DA': '20011114',\n",
       "     'DCOM': '20020402',\n",
       "     'DP': '2001 Sep',\n",
       "     'EDAT': '2001/11/15 10:00',\n",
       "     'IP': '9',\n",
       "     'IS': '0269-2139',\n",
       "     'JID': '8801484',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/03 10:01',\n",
       "     'PG': '691-8',\n",
       "     'PMID': '11707616',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Protein Eng 2001 Sep;14(9):691-8.',\n",
       "     'TA': 'Protein Eng',\n",
       "     'TI': 'Rapid and precise epitope mapping of monoclonal antibodies against Plasmodium falciparum AMA1 by combined phage display of fragments and random peptides.',\n",
       "     'UI': '21564501',\n",
       "     'VI': '14',\n",
       "     'MH': ['Amino Acid Sequence',\n",
       "      'Animal',\n",
       "      'Antibodies, Monoclonal/genetics/*immunology',\n",
       "      'Antibody Specificity',\n",
       "      'Antigens, Protozoan/chemistry/genetics/*immunology',\n",
       "      'Bacteriophages/genetics',\n",
       "      'Disulfides/chemistry',\n",
       "      'Epitope Mapping/*methods',\n",
       "      'Epitopes/immunology',\n",
       "      'Erythrocytes/parasitology',\n",
       "      'Membrane Proteins/chemistry/*immunology',\n",
       "      'Molecular Sequence Data',\n",
       "      'Mutation',\n",
       "      'Peptide Library',\n",
       "      'Peptides/genetics/immunology',\n",
       "      'Plasmodium falciparum/*immunology',\n",
       "      'Protein Binding',\n",
       "      'Protein Structure, Tertiary',\n",
       "      'Protozoan Proteins/chemistry/*immunology',\n",
       "      'Sequence Alignment',\n",
       "      \"Support, Non-U.S. Gov't\"],\n",
       "     'RN': ['0 (Antibodies, Monoclonal)',\n",
       "      '0 (Antigens, Protozoan)',\n",
       "      '0 (Disulfides)',\n",
       "      '0 (Epitopes)',\n",
       "      '0 (Membrane Proteins)',\n",
       "      '0 (Peptide Library)',\n",
       "      '0 (Peptides)',\n",
       "      '0 (Protozoan Proteins)',\n",
       "      '0 (apical membrane antigen I)'],\n",
       "     'FAU': ['Coley, A M',\n",
       "      'Campanale, N V',\n",
       "      'Casey, J L',\n",
       "      'Hodder, A N',\n",
       "      'Crewther, P E',\n",
       "      'Anders, R F',\n",
       "      'Tilley, L M',\n",
       "      'Foley, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '2029',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'Inflammatory cells are thought to be instrumental in the pathophysiology of pulmonary diseases, and control of their recruitment and activation in the lung would appear to be an attractive strategy for therapeutic intervention. Interleukin-8 and related CXC chemokines are involved in the function of neutrophils and T cells, and have been implicated in several lung diseases. Small-molecule antagonists of the interleukin-8 receptors have been identified, which may help elucidate the role of interleukin-8 and related chemokines in the pathophysiology of lung diseases.',\n",
       "     'AD': 'Respiratory, Inflammation and Respiratory Pathogens Center for Excellence in Drug Discovery, GlaxoSmithKline, King of Prussia, PA 19406, USA. douglas_w_hay@sbphrd.com',\n",
       "     'CY': 'England',\n",
       "     'DA': '20011119',\n",
       "     'DCOM': '20020408',\n",
       "     'DP': '2001 Jun',\n",
       "     'EDAT': '2001/11/20 10:00',\n",
       "     'IP': '3',\n",
       "     'IS': '1471-4892',\n",
       "     'JID': '100966133',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/09 10:01',\n",
       "     'PG': '242-7',\n",
       "     'PMID': '11712746',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Review, Tutorial',\n",
       "     'RF': '57',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Curr Opin Pharmacol 2001 Jun;1(3):242-7.',\n",
       "     'TA': 'Curr Opin Pharmacol',\n",
       "     'TI': 'Interleukin-8 receptor antagonists in pulmonary diseases.',\n",
       "     'UI': '21569164',\n",
       "     'VI': '1',\n",
       "     'MH': ['Animal',\n",
       "      'Human',\n",
       "      'Interleukin-8/metabolism/*physiology',\n",
       "      'Lung Diseases/*drug therapy',\n",
       "      'Receptors, Interleukin-8A/*antagonists & inhibitors/physiology'],\n",
       "     'RN': ['0 (Interleukin-8)', '0 (Receptors, Interleukin-8A)'],\n",
       "     'FAU': ['Hay, D W', 'Sarau, H M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '2046',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': 'PROBLEM: It has been suggested that specific mechanisms commonly used by different cellular systems to evade immunologic recognition are involved in the development of endometriosis. To gain insight into this aspect, we looked at the relationship between two of these mechanisms in endometrial stroma and the melanoma system for which the ability to create an environment of immune privilege has been well established. METHOD OF STUDY: Media conditioned by endometrial stromal cultures and malignant melanoma A375 were examined to test their effects on peripheral blood mononuclear cell-mediated cytotoxicity directed against K562 target. Moreover, these media were tested for the concentration of the soluble form of intercellular adhesion molecule-1 (sICAM-1), which has been suggested as a marker for spreading potential. RESULTS: Media conditioned by endometrial stromal cultures exerted a significant suppressive effect on cell cytotoxicity when compared with those derived from malignant melanoma Moreover, the constitutive release of sICAM-1 was significantly higher in supernatants from endometrial stromal than in melanoma cells. CONCLUSIONS: These results indicate that two specific properties suggested to be involved in the ability of tumor cells to evade the immune system are more pronounced in the endometrium than in a malignant melanoma. Since the properties evaluated have been previously demonstrated to be even more notable in endometrial samples derived from endometriosis patients, a role of these mechanisms in the development of the disease may be hypothesized.',\n",
       "     'AD': 'Istituto Scientifico Centro Auxologico Italiano, Milano, Italy. paola.vigano@unimi.it',\n",
       "     'CY': 'Denmark',\n",
       "     'DA': '20011119',\n",
       "     'DCOM': '20020404',\n",
       "     'DP': '2001 Nov',\n",
       "     'EDAT': '2001/11/20 10:00',\n",
       "     'IP': '5',\n",
       "     'IS': '1046-7408',\n",
       "     'JID': '8912860',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/04/18 10:01',\n",
       "     'PG': '342-8',\n",
       "     'PMID': '11712763',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Am J Reprod Immunol 2001 Nov;46(5):342-8.',\n",
       "     'TA': 'Am J Reprod Immunol',\n",
       "     'TI': 'Suppression of natural killer cell function and production of soluble ICAM-1: endometrial stroma versus melanoma.',\n",
       "     'UI': '21569181',\n",
       "     'VI': '46',\n",
       "     'MH': ['Comparative Study',\n",
       "      'Cytotoxicity, Immunologic',\n",
       "      'Endometriosis/etiology',\n",
       "      'Endometrium/*immunology/metabolism',\n",
       "      'Female',\n",
       "      'Human',\n",
       "      '*Immune Tolerance',\n",
       "      'Intercellular Adhesion Molecule-1/*biosynthesis',\n",
       "      'Interleukin-1/pharmacology',\n",
       "      'Killer Cells, Natural/*immunology',\n",
       "      'Melanoma/etiology/*immunology/metabolism',\n",
       "      'Stromal Cells/metabolism',\n",
       "      'Tumor Cells, Cultured'],\n",
       "     'RN': ['0 (Interleukin-1)',\n",
       "      '126547-89-5 (Intercellular Adhesion Molecule-1)'],\n",
       "     'FAU': ['Vigano, P',\n",
       "      'Somigliana, E',\n",
       "      'Di Blasio, A M',\n",
       "      'Cozzolino, S',\n",
       "      'Candiani, M',\n",
       "      'Vignali, M']}},\n",
       "   {'_index': 'genomics-bool',\n",
       "    '_type': '_doc',\n",
       "    '_id': '2420',\n",
       "    '_score': 1.0,\n",
       "    '_source': {'AB': \"In this article we describe detailed pathological and molecular genetics studies in a consanguineous kindred with Pendred's syndrome. The index patient was a 53-year-old female patient with congenital deafness and goiter. Her parents were first-degree cousins. She had a large goiter (150 g) that had been present since childhood. One of her sisters and a niece are also deaf and have goiter as well. The presence of Pendred's syndrome was confirmed by a positive perchlorate test and the demonstration of a Mondini malformation. Thyroid function tests (under levothyroxine [LT4] therapy) were in the euthyroid range with a thyrotropin [TSH] level of 2.8 microU/mL (0.2-3.2), a serum total thyroxine (T4) of 90 nmol/L (54-142), and a serum total triiodothyronine (T3) of 2.7 nmol/L (0.8-2.4). Total thyroidectomy was performed, and the mass in the right lobe was found to have invaded adjacent tissues. The histopathological findings were consistent with a follicular carcinoma with areas of anaplastic transformation and lung metastasis. The patient was treated twice with 100 mCi 131iodine (3,700 MBq) and received suppressive doses of LT4. Postoperatively, the serum thyroglobulin (Tg) levels remained markedly elevated (2,352 to 41,336 ng/mL). The patient died of a sudden severe episode of hemoptysis. Sequence analysis of the PDS gene performed with DNA from the two relatives with Pendred's syndrome revealed the presence of a deletion of thymidine 279 in exon 3, a point mutation that results in a frameshift and a premature stop codon at codon 96 in the pendrin molecule. We concluded that prolonged TSH stimulation because of iodine deficiency or dyshormonogenesis in combination with mutations of oncogenes and/or tumor suppressor genes, may result in the development of follicular thyroid carcinomas that undergo transformation into anaplastic cancers. It is likely that these pathogenetic mechanisms have been involved in the development of aggressive metastatic thyroid cancer in this unusual patient with Pendred's syndrome.\",\n",
       "     'AD': 'Thyroid Unit, Hospital das Clinicas FMUSP, Sao Paulo, Brazil.',\n",
       "     'CY': 'United States',\n",
       "     'DA': '20011121',\n",
       "     'DCOM': '20020522',\n",
       "     'DP': '2001 Oct',\n",
       "     'EDAT': '2001/11/22 10:00',\n",
       "     'IP': '10',\n",
       "     'IS': '1050-7256',\n",
       "     'JID': '9104317',\n",
       "     'LA': 'eng',\n",
       "     'MHDA': '2002/05/25 10:01',\n",
       "     'PG': '981-8',\n",
       "     'PMID': '11716048',\n",
       "     'PST': 'ppublish',\n",
       "     'PT': 'Journal Article',\n",
       "     'SB': 'IM',\n",
       "     'SO': 'Thyroid 2001 Oct;11(10):981-8.',\n",
       "     'TA': 'Thyroid',\n",
       "     'TI': \"Aggressive metastatic follicular thyroid carcinoma with anaplastic transformation arising from a long-standing goiter in a patient with Pendred's syndrome.\",\n",
       "     'UI': '21572133',\n",
       "     'VI': '11',\n",
       "     'MH': ['Adult',\n",
       "      'Amino Acid Sequence',\n",
       "      'Anaplasia',\n",
       "      'Base Sequence',\n",
       "      'Carrier Proteins/genetics',\n",
       "      'Case Report',\n",
       "      'Consanguinity',\n",
       "      'DNA/genetics',\n",
       "      'DNA Mutational Analysis',\n",
       "      'Deafness/*complications/congenital/*genetics',\n",
       "      'Female',\n",
       "      'Goiter/*complications/*genetics',\n",
       "      'Human',\n",
       "      'Lung Neoplasms/secondary',\n",
       "      'Male',\n",
       "      'Middle Age',\n",
       "      'Mutation',\n",
       "      'Pedigree',\n",
       "      \"Support, Non-U.S. Gov't\",\n",
       "      'Syndrome',\n",
       "      'Thyroid Neoplasms/*complications/genetics/pathology'],\n",
       "     'RN': ['0 (Carrier Proteins)',\n",
       "      '0 (sulfate transporter)',\n",
       "      '9007-49-2 (DNA)'],\n",
       "     'FAU': ['Camargo, R',\n",
       "      'Limbert, E',\n",
       "      'Gillam, M',\n",
       "      'Henriques, M M',\n",
       "      'Fernandes, C',\n",
       "      'Catarino, A L',\n",
       "      'Soares, J',\n",
       "      'Alves, V A',\n",
       "      'Kopp, P',\n",
       "      'Medeiros-Neto, G']}}]}}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"molecule\"\n",
    "search_type = \"dfs_query_then_fetch\" # this will use exact document frequencies even for multiple shards\n",
    "body = {\n",
    "  \"query\": {\n",
    "    \"match\" : { \"title-abstract\" : query }\n",
    "  },\n",
    "  \"size\": 10\n",
    "}\n",
    "es.search(index=\"genomics-bool\", search_type=search_type, body=body)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 03.A: _new run and evaluation_\n",
    "Create a new run file (e.g. `boolean.run`), compute the retrieval performance with the function `print_trec_eval()` and compare the results with the baseline run file `baseline.run`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "boolean.run is created\n",
      "Results for boolean.run\n",
      "mean success_at_1              0.05263\n",
      "mean success_at_5              0.1053\n",
      "mean success_at_10             0.1053\n",
      "mean r_precision               0.02801\n",
      "mean precision_at_1            0.05263\n",
      "mean precision_at_5            0.02632\n",
      "mean precision_at_10           0.01316\n",
      "mean precision_at_50           0.0\n",
      "mean precision_at_100          0.0\n",
      "mean precision_at_recall_00    0.05263\n",
      "mean precision_at_recall_01    0.02632\n",
      "mean precision_at_recall_02    0.02632\n",
      "mean precision_at_recall_03    0.02632\n",
      "mean precision_at_recall_04    0.02632\n",
      "mean precision_at_recall_05    0.02632\n",
      "mean precision_at_recall_06    0.0\n",
      "mean precision_at_recall_07    0.0\n",
      "mean precision_at_recall_08    0.0\n",
      "mean precision_at_recall_09    0.0\n",
      "mean precision_at_recall_10    0.0\n",
      "mean average_precision         0.02294\n",
      "Results for baseline.run\n",
      "mean success_at_1              0.1053\n",
      "mean success_at_5              0.2632\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.1803\n",
      "mean precision_at_1            0.1053\n",
      "mean precision_at_5            0.07895\n",
      "mean precision_at_10           0.04737\n",
      "mean precision_at_50           0.0\n",
      "mean precision_at_100          0.0\n",
      "mean precision_at_recall_00    0.1645\n",
      "mean precision_at_recall_01    0.1469\n",
      "mean precision_at_recall_02    0.125\n",
      "mean precision_at_recall_03    0.08114\n",
      "mean precision_at_recall_04    0.07237\n",
      "mean precision_at_recall_05    0.07237\n",
      "mean precision_at_recall_06    0.04605\n",
      "mean precision_at_recall_07    0.04605\n",
      "mean precision_at_recall_08    0.02632\n",
      "mean precision_at_recall_09    0.02632\n",
      "mean precision_at_recall_10    0.02632\n",
      "mean average_precision         0.09836\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# write your code here\n",
    "# BEGIN ANSWER\n",
    "# def bool_trec_run(es, qrels_file, run_file_name, index_name, run_name=\"test_02\", track_scores = True):\n",
    "#     results = es.search(index=index_name, search_type=search_type, body=body)\n",
    "#     count = 1\n",
    "#     qid = 1\n",
    "#     with open(run_file_name, 'w') as run_file:\n",
    "#         for index in results['hits']['hits']:\n",
    "#             id = index['_source']['PMID']\n",
    "#             score = index['_score']\n",
    "#             # for key, value in all_retrieved.items():\n",
    "#             #     if id in value:\n",
    "#             #         print(key)\n",
    "#             run_file.write(str(\"{} Q0 {} {} {} {}\\n\".format(qid, id, count, score, run_name)))\n",
    "#             count += 1\n",
    "    \n",
    "\n",
    "#all_relevant is FIR-s05-training-queries-qrels.txt\n",
    "#all_retrieved is boolean.run\n",
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'boolean.run', \"genomics-bool\")\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'boolean.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'baseline.run')\n",
    "\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 03.B: _Language models_\n",
    "\n",
    "Custom similarities can be configured by tuning the parameters of the built-in similarities. Read more about these (expert) options in the [similarity module](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/index-modules-similarity.html).\n",
    "\n",
    "> Tip: the example similarity settings have to be used in a `\"settings\"` object.\n",
    "> Check your settings and mappings with: `es.indices.get(index='NAME-OF-INDEX')`.\n",
    "\n",
    "__1. Make a run that uses Language Models with [Jelinek-Mercer smoothing](http://lucene.apache.org/core/5_2_1/core/org/apache/lucene/search/similarities/LMJelinekMercerSimilarity.html) (linear interpolation smoothing) on the field `\"all\"` that indexes the fields `\"TI\"` and `\"AB\"`. Use the parameter `lambda=0.2`.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "lmjelinekmercer = {\n",
    "  # BEGIN ANSWER\n",
    "  \"settings\" : {\n",
    "    # a single shard, so we do not suffer from approximate document frequencies\n",
    "    \"index\" : {\n",
    "      \"similarity\" : {\n",
    "        \"my_similarity\" : {\n",
    "          \"type\" : \"LMJelinekMercer\",\n",
    "          \"lambda\" : \"0.2\"\n",
    "        }\n",
    "      }\n",
    "    }  \n",
    "  },\n",
    "  \"mappings\": {\n",
    "      \"properties\": {\n",
    "        \"AB\": {\n",
    "          \"type\": \"text\",\n",
    "          \"similarity\": \"my_similarity\",\n",
    "        },\n",
    "        \"TI\": {\n",
    "          \"type\": \"text\",\n",
    "          \"similarity\": \"my_similarity\",\n",
    "        },\n",
    "      }\n",
    "  }\n",
    "  # END ANSWER\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_364/1573036711.py:24: DeprecationWarning: The 'body' parameter is deprecated for the 'create' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  es.indices.create(index=index_name, body=body)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [42], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mindex_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdata01/FIR-s05-medline.json\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgenomics-jm\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlmjelinekmercer\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn [23], line 29\u001b[0m, in \u001b[0;36mindex_documents\u001b[0;34m(es, collection_file_name, index_name, body)\u001b[0m\n\u001b[1;32m     27\u001b[0m create_index(es, index_name, body)\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# bulk index the documents from file_name\u001b[39;00m\n\u001b[0;32m---> 29\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43melasticsearch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhelpers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbulk\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     30\u001b[0m \u001b[43m    \u001b[49m\u001b[43mes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[43m    \u001b[49m\u001b[43mread_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcollection_file_name\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunk_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2000\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrequest_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m30\u001b[39;49m\n\u001b[1;32m     35\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/helpers/actions.py:410\u001b[0m, in \u001b[0;36mbulk\u001b[0;34m(client, actions, stats_only, ignore_status, *args, **kwargs)\u001b[0m\n\u001b[1;32m    408\u001b[0m \u001b[39m# make streaming_bulk yield successful results so we can count them\u001b[39;00m\n\u001b[1;32m    409\u001b[0m kwargs[\u001b[39m\"\u001b[39m\u001b[39myield_ok\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[0;32m--> 410\u001b[0m \u001b[39mfor\u001b[39;00m ok, item \u001b[39min\u001b[39;00m streaming_bulk(\n\u001b[1;32m    411\u001b[0m     client, actions, ignore_status\u001b[39m=\u001b[39mignore_status, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs\n\u001b[1;32m    412\u001b[0m ):\n\u001b[1;32m    413\u001b[0m     \u001b[39m# go through request-response pairs and detect failures\u001b[39;00m\n\u001b[1;32m    414\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m ok:\n\u001b[1;32m    415\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m stats_only:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/helpers/actions.py:329\u001b[0m, in \u001b[0;36mstreaming_bulk\u001b[0;34m(client, actions, chunk_size, max_chunk_bytes, raise_on_error, expand_action_callback, raise_on_exception, max_retries, initial_backoff, max_backoff, yield_ok, ignore_status, *args, **kwargs)\u001b[0m\n\u001b[1;32m    326\u001b[0m     time\u001b[39m.\u001b[39msleep(\u001b[39mmin\u001b[39m(max_backoff, initial_backoff \u001b[39m*\u001b[39m \u001b[39m2\u001b[39m \u001b[39m*\u001b[39m\u001b[39m*\u001b[39m (attempt \u001b[39m-\u001b[39m \u001b[39m1\u001b[39m)))\n\u001b[1;32m    328\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 329\u001b[0m     \u001b[39mfor\u001b[39;00m data, (ok, info) \u001b[39min\u001b[39;00m \u001b[39mzip\u001b[39m(\n\u001b[1;32m    330\u001b[0m         bulk_data,\n\u001b[1;32m    331\u001b[0m         _process_bulk_chunk(\n\u001b[1;32m    332\u001b[0m             client,\n\u001b[1;32m    333\u001b[0m             bulk_actions,\n\u001b[1;32m    334\u001b[0m             bulk_data,\n\u001b[1;32m    335\u001b[0m             raise_on_exception,\n\u001b[1;32m    336\u001b[0m             raise_on_error,\n\u001b[1;32m    337\u001b[0m             ignore_status,\n\u001b[1;32m    338\u001b[0m             \u001b[39m*\u001b[39margs,\n\u001b[1;32m    339\u001b[0m             \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs\n\u001b[1;32m    340\u001b[0m         ),\n\u001b[1;32m    341\u001b[0m     ):\n\u001b[1;32m    343\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m ok:\n\u001b[1;32m    344\u001b[0m             action, info \u001b[39m=\u001b[39m info\u001b[39m.\u001b[39mpopitem()\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/helpers/actions.py:240\u001b[0m, in \u001b[0;36m_process_bulk_chunk\u001b[0;34m(client, bulk_actions, bulk_data, raise_on_exception, raise_on_error, ignore_status, *args, **kwargs)\u001b[0m\n\u001b[1;32m    236\u001b[0m     ignore_status \u001b[39m=\u001b[39m (ignore_status,)\n\u001b[1;32m    238\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    239\u001b[0m     \u001b[39m# send the actual request\u001b[39;00m\n\u001b[0;32m--> 240\u001b[0m     resp \u001b[39m=\u001b[39m client\u001b[39m.\u001b[39;49mbulk(\u001b[39m*\u001b[39;49margs, body\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\\n\u001b[39;49;00m\u001b[39m\"\u001b[39;49m\u001b[39m.\u001b[39;49mjoin(bulk_actions) \u001b[39m+\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39m\\n\u001b[39;49;00m\u001b[39m\"\u001b[39;49m, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    241\u001b[0m \u001b[39mexcept\u001b[39;00m TransportError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    242\u001b[0m     gen \u001b[39m=\u001b[39m _process_bulk_chunk_error(\n\u001b[1;32m    243\u001b[0m         error\u001b[39m=\u001b[39me,\n\u001b[1;32m    244\u001b[0m         bulk_data\u001b[39m=\u001b[39mbulk_data,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    247\u001b[0m         raise_on_error\u001b[39m=\u001b[39mraise_on_error,\n\u001b[1;32m    248\u001b[0m     )\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/client/utils.py:347\u001b[0m, in \u001b[0;36mquery_params.<locals>._wrapper.<locals>._wrapped\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    345\u001b[0m     \u001b[39mif\u001b[39;00m p \u001b[39min\u001b[39;00m kwargs:\n\u001b[1;32m    346\u001b[0m         params[p] \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(p)\n\u001b[0;32m--> 347\u001b[0m \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, params\u001b[39m=\u001b[39;49mparams, headers\u001b[39m=\u001b[39;49mheaders, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/client/__init__.py:472\u001b[0m, in \u001b[0;36mElasticsearch.bulk\u001b[0;34m(self, body, index, doc_type, params, headers)\u001b[0m\n\u001b[1;32m    469\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mEmpty value passed for a required argument \u001b[39m\u001b[39m'\u001b[39m\u001b[39mbody\u001b[39m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    471\u001b[0m body \u001b[39m=\u001b[39m _bulk_body(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtransport\u001b[39m.\u001b[39mserializer, body)\n\u001b[0;32m--> 472\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mtransport\u001b[39m.\u001b[39;49mperform_request(\n\u001b[1;32m    473\u001b[0m     \u001b[39m\"\u001b[39;49m\u001b[39mPOST\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    474\u001b[0m     _make_path(index, doc_type, \u001b[39m\"\u001b[39;49m\u001b[39m_bulk\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[1;32m    475\u001b[0m     params\u001b[39m=\u001b[39;49mparams,\n\u001b[1;32m    476\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    477\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    478\u001b[0m )\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/transport.py:427\u001b[0m, in \u001b[0;36mTransport.perform_request\u001b[0;34m(self, method, url, headers, params, body)\u001b[0m\n\u001b[1;32m    424\u001b[0m connection \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_connection()\n\u001b[1;32m    426\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 427\u001b[0m     status, headers_response, data \u001b[39m=\u001b[39m connection\u001b[39m.\u001b[39;49mperform_request(\n\u001b[1;32m    428\u001b[0m         method,\n\u001b[1;32m    429\u001b[0m         url,\n\u001b[1;32m    430\u001b[0m         params,\n\u001b[1;32m    431\u001b[0m         body,\n\u001b[1;32m    432\u001b[0m         headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    433\u001b[0m         ignore\u001b[39m=\u001b[39;49mignore,\n\u001b[1;32m    434\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[1;32m    435\u001b[0m     )\n\u001b[1;32m    437\u001b[0m     \u001b[39m# Lowercase all the header names for consistency in accessing them.\u001b[39;00m\n\u001b[1;32m    438\u001b[0m     headers_response \u001b[39m=\u001b[39m {\n\u001b[1;32m    439\u001b[0m         header\u001b[39m.\u001b[39mlower(): value \u001b[39mfor\u001b[39;00m header, value \u001b[39min\u001b[39;00m headers_response\u001b[39m.\u001b[39mitems()\n\u001b[1;32m    440\u001b[0m     }\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/elasticsearch/connection/http_urllib3.py:255\u001b[0m, in \u001b[0;36mUrllib3HttpConnection.perform_request\u001b[0;34m(self, method, url, params, body, timeout, ignore, headers)\u001b[0m\n\u001b[1;32m    252\u001b[0m     body \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_gzip_compress(body)\n\u001b[1;32m    253\u001b[0m     request_headers[\u001b[39m\"\u001b[39m\u001b[39mcontent-encoding\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mgzip\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 255\u001b[0m response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpool\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    256\u001b[0m     method, url, body, retries\u001b[39m=\u001b[39;49mRetry(\u001b[39mFalse\u001b[39;49;00m), headers\u001b[39m=\u001b[39;49mrequest_headers, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkw\n\u001b[1;32m    257\u001b[0m )\n\u001b[1;32m    258\u001b[0m response_headers \u001b[39m=\u001b[39m {\n\u001b[1;32m    259\u001b[0m     header\u001b[39m.\u001b[39mlower(): value \u001b[39mfor\u001b[39;00m header, value \u001b[39min\u001b[39;00m response\u001b[39m.\u001b[39mheaders\u001b[39m.\u001b[39mitems()\n\u001b[1;32m    260\u001b[0m }\n\u001b[1;32m    261\u001b[0m duration \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mtime() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:665\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    662\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_prepare_proxy(conn)\n\u001b[1;32m    664\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 665\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    666\u001b[0m     conn,\n\u001b[1;32m    667\u001b[0m     method,\n\u001b[1;32m    668\u001b[0m     url,\n\u001b[1;32m    669\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    670\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    671\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    672\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    673\u001b[0m )\n\u001b[1;32m    675\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    676\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    677\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    678\u001b[0m \u001b[39m# mess.\u001b[39;00m\n\u001b[1;32m    679\u001b[0m response_conn \u001b[39m=\u001b[39m conn \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m release_conn \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:421\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    416\u001b[0m             httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39mgetresponse()\n\u001b[1;32m    417\u001b[0m         \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    418\u001b[0m             \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    419\u001b[0m             \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    420\u001b[0m             \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[0;32m--> 421\u001b[0m             six\u001b[39m.\u001b[39;49mraise_from(e, \u001b[39mNone\u001b[39;49;00m)\n\u001b[1;32m    422\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    423\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_raise_timeout(err\u001b[39m=\u001b[39me, url\u001b[39m=\u001b[39murl, timeout_value\u001b[39m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:416\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    413\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[1;32m    414\u001b[0m     \u001b[39m# Python 3\u001b[39;00m\n\u001b[1;32m    415\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 416\u001b[0m         httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    417\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    418\u001b[0m         \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    419\u001b[0m         \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    420\u001b[0m         \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[1;32m    421\u001b[0m         six\u001b[39m.\u001b[39mraise_from(e, \u001b[39mNone\u001b[39;00m)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1374\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1372\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m   1373\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1374\u001b[0m         response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1375\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n\u001b[1;32m   1376\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mclose()\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[39m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    319\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n\u001b[1;32m    320\u001b[0m         \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_read_status\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfp\u001b[39m.\u001b[39;49mreadline(_MAXLINE \u001b[39m+\u001b[39;49m \u001b[39m1\u001b[39;49m), \u001b[39m\"\u001b[39m\u001b[39miso-8859-1\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(line) \u001b[39m>\u001b[39m _MAXLINE:\n\u001b[1;32m    281\u001b[0m         \u001b[39mraise\u001b[39;00m LineTooLong(\u001b[39m\"\u001b[39m\u001b[39mstatus line\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m    704\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sock\u001b[39m.\u001b[39;49mrecv_into(b)\n\u001b[1;32m    706\u001b[0m     \u001b[39mexcept\u001b[39;00m timeout:\n\u001b[1;32m    707\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_timeout_occurred \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-jm', body=lmjelinekmercer)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lmjelinekmercer.run is created\n"
     ]
    }
   ],
   "source": [
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'lmjelinekmercer.run', 'genomics-jm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__2. Make a run that uses Language Models with [Dirichelet smoothing](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-similarity.html) to index the fields `\"TI\"` and `\"AB\"`. Use the parameter `mu=2000`.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "dirichlet = {\n",
    "  # BEGIN ANSWER\n",
    "  \"settings\" : {\n",
    "    # a single shard, so we do not suffer from approximate document frequencies\n",
    "    \"index\" : {\n",
    "      \"similarity\" : {\n",
    "        \"my_similarity\" : {\n",
    "          \"type\" : \"LMDirichlet\",\n",
    "          \"mu\" : \"2000\"\n",
    "        }\n",
    "      }\n",
    "    }  \n",
    "  },\n",
    "  \"mappings\": {\n",
    "      \"properties\": {\n",
    "        \"AB\": {\n",
    "          \"type\": \"text\",\n",
    "          \"similarity\": \"my_similarity\",\n",
    "        },\n",
    "        \"TI\": {\n",
    "          \"type\": \"text\",\n",
    "          \"similarity\": \"my_similarity\",\n",
    "        },\n",
    "      }\n",
    "  }\n",
    "  # END ANSWER\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_799/1573036711.py:24: DeprecationWarning: The 'body' parameter is deprecated for the 'create' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  es.indices.create(index=index_name, body=body)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(263080, [])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-dirichlet', body=dirichlet)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dirichlet.run is created\n"
     ]
    }
   ],
   "source": [
    "make_trec_run(es, 'data01/FIR-s05-training-queries-simple.txt', 'dirichlet.run', 'genomics-dirichlet')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 03.C: _Model comparison_\n",
    "\n",
    "\n",
    "__1. Compute the performance results of the `lmjelinekmercer.run` and `dirichelet.run`. Compare them with those of the `baseline.run` and `boolean.run`. Evaluate the runs using the `print_trec_eval` function. Performing statistical tests may help strengthen your claims.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for boolean.run\n",
      "mean success_at_1              0.05263\n",
      "mean success_at_5              0.1053\n",
      "mean success_at_10             0.1053\n",
      "mean r_precision               0.02801\n",
      "mean precision_at_1            0.05263\n",
      "mean precision_at_5            0.02632\n",
      "mean precision_at_10           0.01316\n",
      "mean precision_at_50           0.0\n",
      "mean precision_at_100          0.0\n",
      "mean precision_at_recall_00    0.05263\n",
      "mean precision_at_recall_01    0.02632\n",
      "mean precision_at_recall_02    0.02632\n",
      "mean precision_at_recall_03    0.02632\n",
      "mean precision_at_recall_04    0.02632\n",
      "mean precision_at_recall_05    0.02632\n",
      "mean precision_at_recall_06    0.0\n",
      "mean precision_at_recall_07    0.0\n",
      "mean precision_at_recall_08    0.0\n",
      "mean precision_at_recall_09    0.0\n",
      "mean precision_at_recall_10    0.0\n",
      "mean average_precision         0.02294\n",
      "Results for baseline.run\n",
      "mean success_at_1              0.1053\n",
      "mean success_at_5              0.2632\n",
      "mean success_at_10             0.3158\n",
      "mean r_precision               0.1803\n",
      "mean precision_at_1            0.1053\n",
      "mean precision_at_5            0.07895\n",
      "mean precision_at_10           0.04737\n",
      "mean precision_at_50           0.0\n",
      "mean precision_at_100          0.0\n",
      "mean precision_at_recall_00    0.1645\n",
      "mean precision_at_recall_01    0.1469\n",
      "mean precision_at_recall_02    0.125\n",
      "mean precision_at_recall_03    0.08114\n",
      "mean precision_at_recall_04    0.07237\n",
      "mean precision_at_recall_05    0.07237\n",
      "mean precision_at_recall_06    0.04605\n",
      "mean precision_at_recall_07    0.04605\n",
      "mean precision_at_recall_08    0.02632\n",
      "mean precision_at_recall_09    0.02632\n",
      "mean precision_at_recall_10    0.02632\n",
      "mean average_precision         0.09836\n",
      "Results for lmjelinekmercer.run\n",
      "mean success_at_1              0.1316\n",
      "mean success_at_5              0.2895\n",
      "mean success_at_10             0.2895\n",
      "mean r_precision               0.2154\n",
      "mean precision_at_1            0.1316\n",
      "mean precision_at_5            0.08947\n",
      "mean precision_at_10           0.05263\n",
      "mean precision_at_50           0.0\n",
      "mean precision_at_100          0.0\n",
      "mean precision_at_recall_00    0.1908\n",
      "mean precision_at_recall_01    0.182\n",
      "mean precision_at_recall_02    0.182\n",
      "mean precision_at_recall_03    0.1645\n",
      "mean precision_at_recall_04    0.1645\n",
      "mean precision_at_recall_05    0.1645\n",
      "mean precision_at_recall_06    0.09868\n",
      "mean precision_at_recall_07    0.09868\n",
      "mean precision_at_recall_08    0.07895\n",
      "mean precision_at_recall_09    0.07895\n",
      "mean precision_at_recall_10    0.07895\n",
      "mean average_precision         0.1481\n",
      "Results for dirichlet.run\n",
      "mean success_at_1              0.05263\n",
      "mean success_at_5              0.2632\n",
      "mean success_at_10             0.2632\n",
      "mean r_precision               0.1408\n",
      "mean precision_at_1            0.05263\n",
      "mean precision_at_5            0.08947\n",
      "mean precision_at_10           0.04474\n",
      "mean precision_at_50           0.0\n",
      "mean precision_at_100          0.0\n",
      "mean precision_at_recall_00    0.1237\n",
      "mean precision_at_recall_01    0.09737\n",
      "mean precision_at_recall_02    0.09211\n",
      "mean precision_at_recall_03    0.07456\n",
      "mean precision_at_recall_04    0.06579\n",
      "mean precision_at_recall_05    0.06579\n",
      "mean precision_at_recall_06    0.0\n",
      "mean precision_at_recall_07    0.0\n",
      "mean precision_at_recall_08    0.0\n",
      "mean precision_at_recall_09    0.0\n",
      "mean precision_at_recall_10    0.0\n",
      "mean average_precision         0.05731\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "# your comments here\n",
    "# BEGIN ANSWER\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'boolean.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'baseline.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'lmjelinekmercer.run')\n",
    "print_trec_eval('data01/FIR-s05-training-qrels.txt', 'dirichlet.run')\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top20 retrieved documents baseline.run\n",
      "1 Q0 11929828 1 43.560047 test\n",
      "1 Q0 11751903 2 43.560047 test\n",
      "1 Q0 12384701 3 43.560047 test\n",
      "1 Q0 12065641 4 43.560047 test\n",
      "1 Q0 11980715 5 43.560047 test\n",
      "1 Q0 12126481 6 43.560047 test\n",
      "1 Q0 12455049 7 43.560047 test\n",
      "1 Q0 12444545 8 43.560047 test\n",
      "1 Q0 12431783 9 43.560047 test\n",
      "1 Q0 12204896 10 43.560047 test\n",
      "3 Q0 12167152 1 27.794035 test\n",
      "3 Q0 12435397 2 27.794035 test\n",
      "3 Q0 12370291 3 27.794035 test\n",
      "3 Q0 12036890 4 27.794035 test\n",
      "3 Q0 11761710 5 27.794035 test\n",
      "3 Q0 12645611 6 27.794035 test\n",
      "3 Q0 11937514 7 27.794035 test\n",
      "3 Q0 12018119 8 27.794035 test\n",
      "3 Q0 12527194 9 27.794035 test\n",
      "3 Q0 12148460 10 27.794035 test\n",
      "\n",
      "Top20 retrieved documents boolean.run\n",
      "1 Q0 11762751 1 7.0 test\n",
      "1 Q0 11767002 2 7.0 test\n",
      "1 Q0 11809764 3 7.0 test\n",
      "1 Q0 11724777 4 7.0 test\n",
      "1 Q0 11751903 5 7.0 test\n",
      "1 Q0 11783178 6 7.0 test\n",
      "1 Q0 11788578 7 7.0 test\n",
      "1 Q0 11795494 8 7.0 test\n",
      "1 Q0 11732804 9 7.0 test\n",
      "1 Q0 11755166 10 7.0 test\n",
      "3 Q0 11759083 1 4.0 test\n",
      "3 Q0 11761710 2 4.0 test\n",
      "3 Q0 11770079 3 4.0 test\n",
      "3 Q0 11793389 4 4.0 test\n",
      "3 Q0 11793697 5 4.0 test\n",
      "3 Q0 11793698 6 4.0 test\n",
      "3 Q0 11796228 7 4.0 test\n",
      "3 Q0 11800169 8 4.0 test\n",
      "3 Q0 11805739 9 4.0 test\n",
      "3 Q0 11814575 10 4.0 test\n",
      "\n",
      "Top20 retrieved documents lmjelinekmercer.run\n",
      "1 Q0 12368211 1 44.58327 test\n",
      "1 Q0 11929828 2 44.58327 test\n",
      "1 Q0 11980715 3 44.58327 test\n",
      "1 Q0 11882322 4 44.58327 test\n",
      "1 Q0 12065641 5 44.58327 test\n",
      "1 Q0 12384701 6 44.58327 test\n",
      "1 Q0 12431783 7 44.58327 test\n",
      "1 Q0 12455049 8 44.58327 test\n",
      "1 Q0 12495933 9 44.58327 test\n",
      "1 Q0 12126481 10 44.58327 test\n",
      "3 Q0 12527194 1 29.90406 test\n",
      "3 Q0 11993511 2 29.90406 test\n",
      "3 Q0 12499268 3 29.90406 test\n",
      "3 Q0 12167152 4 29.90406 test\n",
      "3 Q0 11937514 5 29.90406 test\n",
      "3 Q0 12234520 6 29.90406 test\n",
      "3 Q0 12135477 7 29.90406 test\n",
      "3 Q0 12468090 8 29.90406 test\n",
      "3 Q0 12370291 9 29.90406 test\n",
      "3 Q0 12645611 10 29.90406 test\n",
      "\n",
      "Top20 retrieved documents dirichlet.run\n",
      "1 Q0 11751903 1 16.14804 test\n",
      "1 Q0 12112322 2 16.14804 test\n",
      "1 Q0 12065641 3 16.14804 test\n",
      "1 Q0 12126481 4 16.14804 test\n",
      "1 Q0 12384701 5 16.14804 test\n",
      "1 Q0 12242284 6 16.14804 test\n",
      "1 Q0 12431783 7 16.14804 test\n",
      "1 Q0 11827966 8 16.14804 test\n",
      "1 Q0 12444545 9 16.14804 test\n",
      "1 Q0 12204896 10 16.14804 test\n",
      "3 Q0 12527194 1 14.332972 test\n",
      "3 Q0 11993511 2 14.332972 test\n",
      "3 Q0 11937514 3 14.332972 test\n",
      "3 Q0 12018119 4 14.332972 test\n",
      "3 Q0 12435397 5 14.332972 test\n",
      "3 Q0 12370291 6 14.332972 test\n",
      "3 Q0 12234520 7 14.332972 test\n",
      "3 Q0 12028792 8 14.332972 test\n",
      "3 Q0 12499268 9 14.332972 test\n",
      "3 Q0 12231201 10 14.332972 test\n"
     ]
    }
   ],
   "source": [
    "print('Top20 retrieved documents baseline.run')\n",
    "! head -20 baseline.run\n",
    "\n",
    "print('\\nTop20 retrieved documents boolean.run')\n",
    "! head -20 boolean.run\n",
    "\n",
    "print('\\nTop20 retrieved documents lmjelinekmercer.run')\n",
    "! head -20 lmjelinekmercer.run\n",
    "\n",
    "print('\\nTop20 retrieved documents dirichlet.run')\n",
    "! head -20 dirichlet.run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "__2. Provide below your comments and interpretations of the results. Why, in your opinion, one model of search is better than the others?__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# answer as a comment:\n",
    "# One possibility is definitely the way the documents are indexed and mapped. One index may be mapped better than the other and \n",
    "# as a result, the documents are retrieved in a different order. Another possibility is the way the queries are processed.\n",
    "# jelinekmercer and dirichlet are both probabilistic models, so they may be more accurate than the boolean model.\n",
    "# however the boolean model is much faster than the other two models, so it may be more useful for large datasets.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 04: Index improvements: Tokenization\n",
    "<span style=\"background:red; color: white;\">__You are advised to work on this part after Lecture 03__</span>\n",
    "\n",
    "\n",
    "\n",
    "## Background\n",
    "The following part of the assignment requires some self-study of the ElasticSearch tools to support the improvemnet of the indexing. Please read the:\n",
    "* [Index Settings and Mappings](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/indices-create-index.html).\n",
    "* Elasticsearch [Analyzers](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/analysis.html) contain many options for improving your search engine.\n",
    "\n",
    "> You are suggested to use the [Python Elasticsearch Client](https://elasticsearch-py.readthedocs.io) library documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example: _ElasticSearch Analyzers for tokenization_\n",
    "\n",
    "The amount and quality of the tokens used to construct the inverted index are of great importance. In ElasticSearch, mappings and settings also allow specifying what [Analyzer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis.html) is used to tokenize your documents and queries. In the mappings below, use the _Dutch_ analyzer for the field `\"all\"`):\n",
    "\n",
    "> Usually, the same analyzer should be applied to documents and queries, but \n",
    "> Elasticsearch allows you to specify a `\"search_analyzer\"` that is used on \n",
    "> your queries (which we do not need to use in the assignment)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jarron/.local/lib/python3.10/site-packages/elasticsearch/connection/base.py:200: ElasticsearchWarning: Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.17/security-minimal-setup.html to enable security.\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n",
      "/tmp/ipykernel_364/1573036711.py:24: DeprecationWarning: The 'body' parameter is deprecated for the 'create' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  es.indices.create(index=index_name, body=body)\n"
     ]
    }
   ],
   "source": [
    "analyzer_test = {\n",
    "  \"mappings\": {\n",
    "      \"properties\": {\n",
    "        \"all\": {\n",
    "          \"type\": \"text\",\n",
    "          \"analyzer\": \"dutch\"\n",
    "        }\n",
    "      }\n",
    "  }\n",
    "}\n",
    "\n",
    "# create the index, but don't index any documents:\n",
    "create_index(es, 'test-tokens', body=analyzer_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The analyzer defined for the `\"all\"` field can be tested [as follows](https://elasticsearch-py.readthedocs.io/en/master/api.html#indices). Translated to English the text says: _\"This is a Dutch sentence\"_. \n",
    "\n",
    "> The following script identifies the tokens (based on the use of the dutch tokenizer): try with different tokenizers and different sentences to see how the tokens are created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'tokens': [{'end_offset': 20,\n",
      "             'position': 2,\n",
      "             'start_offset': 9,\n",
      "             'token': 'nederland',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 27,\n",
      "             'position': 3,\n",
      "             'start_offset': 21,\n",
      "             'token': 'zinn',\n",
      "             'type': '<ALPHANUM>'}]}\n"
     ]
    }
   ],
   "source": [
    "from pprint import pprint # pretty print\n",
    "\n",
    "body = { \"field\": \"all\", \"text\": \"dit zijn nederlandse zinnen\"}\n",
    "tokens = es.indices.analyze(index='test-tokens', body=body)\n",
    "pprint(tokens)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Exercise 04.A: _chat language analyzer_\n",
    "\n",
    "Read the documentation for [Custom Analyzer](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/analysis-custom-analyzer.html). \n",
    "Make a custom analyzer for _English chat language_. The analyzer should do the following:\n",
    "* change common abbreviations to the full forms: \n",
    "  * _b4_ to _before_, \n",
    "  * _abt_ to _about_, \n",
    "  * _chk_ to _check_, \n",
    "  * _dm_ to _direct message_,\n",
    "  * _f2f_ to _face-to-face_\n",
    "* use the _standard_ tokenizer;\n",
    "* put everything to lower-case;\n",
    "* filter English stopwords."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_364/1573036711.py:24: DeprecationWarning: The 'body' parameter is deprecated for the 'create' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  es.indices.create(index=index_name, body=body)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'tokens': [{'end_offset': 4,\n",
      "             'position': 0,\n",
      "             'start_offset': 0,\n",
      "             'token': 'done',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 10,\n",
      "             'position': 2,\n",
      "             'start_offset': 8,\n",
      "             'token': 'before',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 16,\n",
      "             'position': 3,\n",
      "             'start_offset': 12,\n",
      "             'token': 'what',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 20,\n",
      "             'position': 4,\n",
      "             'start_offset': 17,\n",
      "             'token': 'about',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 22,\n",
      "             'position': 5,\n",
      "             'start_offset': 21,\n",
      "             'token': 'direct',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 23,\n",
      "             'position': 6,\n",
      "             'start_offset': 22,\n",
      "             'token': 'message',\n",
      "             'type': '<ALPHANUM>'},\n",
      "            {'end_offset': 26,\n",
      "             'position': 7,\n",
      "             'start_offset': 24,\n",
      "             'token': 'me',\n",
      "             'type': '<ALPHANUM>'}]}\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "tweet_analyzer = {\n",
    "  # BEGIN ANSWER\n",
    "  \"settings\": {\n",
    "    \"analysis\": {\n",
    "      \"analyzer\": {\n",
    "        \"my_analyzer\": {\n",
    "          \"type\": \"custom\",\n",
    "          \"tokenizer\": \"standard\",\n",
    "          \"char_filter\": [\n",
    "              \"char_mapping\"\n",
    "            ],\n",
    "            \"filter\": [\n",
    "              \"lowercase\",\n",
    "              \"stop_filter\"\n",
    "            ]\n",
    "        }\n",
    "      },\n",
    "      \"char_filter\": {\n",
    "          \"char_mapping\": {\n",
    "            \"type\": \"mapping\",\n",
    "            \"mappings\": [\n",
    "              \"b4 => before\",\n",
    "              \"abt => about\",\n",
    "              \"chk => check\",\n",
    "              \"dm => direct message\",\n",
    "              \"f2f => face to face\"\n",
    "            ]\n",
    "          }\n",
    "        },\n",
    "        \"filter\": {\n",
    "          \"stop_filter\": {\n",
    "                \"type\" : \"stop\",\n",
    "                \"stopwords\" : \"_english_\"\n",
    "              }\n",
    "        }\n",
    "    }\n",
    "  },\n",
    "  \"mappings\": {\n",
    "    \"properties\": {\n",
    "      \"all\" : {\n",
    "        \"type\" : \"text\",\n",
    "        \"analyzer\": \"my_analyzer\"\n",
    "      },\n",
    "    }\n",
    "  }\n",
    "  # END ANSWER\n",
    "}\n",
    "\n",
    "# create the index, but don't index any documents:\n",
    "create_index(es, 'genomics', body=tweet_analyzer)\n",
    "body = { \"field\": \"all\", \"text\": \"done it b4! what abt dm me?\"}\n",
    "tokens = es.indices.analyze(index='genomics', body=body)\n",
    "pprint(tokens)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 04.B: Stemmers\n",
    "\n",
    "Referring at Exercise 02.A, we have seen that queries like `molecule` and `molecular` retrieve different sets of documents. Lemmatizer and stemmers can help the indexing and search of 'similar' terms, and retrieve more consistent sets of documents.\n",
    "\n",
    "__Use the ElasticSearch [Stemming](https://www.elastic.co/guide/en/elasticsearch/reference/current/stemming.html) to index the document collection. Then retrieve documents with the queries `molecule` and `molecular` and comment on the eventual differences with the previous query results.__\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "body = {\n",
    "  # BEGIN ANSWER\n",
    "  \"settings\": {\n",
    "    \"analysis\": {\n",
    "      \"analyzer\": {\n",
    "        \"my_analyzer\": {\n",
    "          \"tokenizer\": \"whitespace\",\n",
    "          \"filter\": [ \"stemmer\" ]\n",
    "        }\n",
    "      },\n",
    "      \"analyzer\": {\n",
    "        \"tags_analyzer\": {\n",
    "          \"type\": \"custom\",\n",
    "          \"filter\": [\n",
    "            \"standard\",\n",
    "            \"lowercase\",\n",
    "            \"filter_stemmer\"\n",
    "          ],\n",
    "          \"tokenizer\": \"standard\"\n",
    "        }\n",
    "      }\n",
    "    },\n",
    "    \"mappings\": {\n",
    "        \"properties\": {\n",
    "            \"analyzer\": \"tags_analyzer\",\n",
    "            \"type\": \"text\"\n",
    "        }\n",
    "      }\n",
    "  }\n",
    "  # END ANSWER\n",
    "}\n",
    "\n",
    "body = {\n",
    "  # BEGIN ANSWER\n",
    "  # END ANSWER\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_364/1573036711.py:24: DeprecationWarning: The 'body' parameter is deprecated for the 'create' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  es.indices.create(index=index_name, body=body)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(263080, [])"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Connect to the ElasticSearch server\n",
    "es = elasticsearch.Elasticsearch(host='localhost')  # in case you use Docker, the host is 'elasticsearch'\n",
    "\n",
    "# Index the collection into the index called 'genomics-stem'\n",
    "index_documents(es, 'data01/FIR-s05-medline.json', 'genomics-stem', body)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Retrieve documents with the queries `molecule` and `molecular` and comment on the eventual differences with the previous query results.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jarron/.local/lib/python3.10/site-packages/elasticsearch/connection/base.py:200: ElasticsearchWarning: Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.17/security-minimal-setup.html to enable security.\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n",
      "/home/jarron/.local/lib/python3.10/site-packages/elasticsearch/connection/base.py:200: ElasticsearchWarning: this request accesses system indices: [.apm-agent-configuration, .apm-custom-link, .kibana_7.17.6_001, .kibana_task_manager_7.17.6_001, .tasks], but in a future major version, direct access to system indices will be prevented by default\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11150\n",
      "4354\n"
     ]
    }
   ],
   "source": [
    "#THIS IS GRADED!\n",
    "# BEGIN ANSWER\n",
    "s1 = Search(using=es)\n",
    "q = Q(\"multi_match\", query=\"molecule\")\n",
    "s1 = s1.query(q)\n",
    "print(s1.count())\n",
    "\n",
    "s2 = Search(using=es)\n",
    "q = Q(\"multi_match\", query=\"molecular\")\n",
    "s2 = s1.query(q)\n",
    "print(s2.count())\n",
    "# END ANSWER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comment here about the eventual different results you get \n",
    "# -> words that are stemmed, like molecule and molecules, should improve the retrieval results (in terms of amount of retrieved dos)\n",
    "# -> words that are not stemmed, like 'molecular' should not see much differernt results\n",
    "\n",
    "# the point of the exercise is not to have the same results with molecule and molecular, but \n",
    "# what the stemmer does, and reason after that."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BONUS PART: _Implement your own similarity measure_ \n",
    "\n",
    "We have only seen the results of using the analyzer to queries. The analyzer results from the _documents_ are available using the `termvectors()` function, as follows for document `id=3`: (Additionally, we can get overall field statistics, such as the number of documents)\n",
    "\n",
    "> First, index the collection again. While waiting, have a coffee or tea :) \n",
    "\n",
    "> `id=3` refers to the internal document identifiers, so not to the Pubmed identifier.\n",
    "\n",
    "_The bonus exercise is not mandatory. It can compensate for lower grades of other exercises._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import elasticsearch\n",
    "es = elasticsearch.Elasticsearch(host='localhost')\n",
    "\n",
    "# index_documents(es, 'data/FIR-05-medline.json', 'genomics-base')\n",
    "\n",
    "es.termvectors(index=\"genomics-base\", id=\"3\", fields=\"TI\", \n",
    "               term_statistics=True, field_statistics=True, offsets=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Implement the BM25 similarity\n",
    "\n",
    "Complete the function `bm25_similarity()` below by implementing the BM25 similarity as described by in Section 11.4.3 of [Manning, Raghavan and Schuetze, Chapter 11](https://nlp.stanford.edu/IR-book/pdf/11prob.pdf). Are you able to replicate the score of ElasitcSearch (9.55)? If not, are you using a different variant of the BM25 model? Provide your comments in plain text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS GRADED!\n",
    "\n",
    "import math\n",
    "\n",
    "# math.log(x) computes the logarithm of x\n",
    "\n",
    "def bm25_similarity (query, doc_id):\n",
    "\n",
    "    # Get the query tokens (see above)\n",
    "    query_tokens = es.indices.analyze(index='genomics-base', body={\"field\":\"TI\", \"text\": query})\n",
    "    tokens = query_tokens['tokens']\n",
    "\n",
    "    # Get the term vector for doc_id and the field statistics\n",
    "    term_vector = es.termvectors(index=\"genomics-base\", id=doc_id, fields=\"TI\", \n",
    "                  term_statistics=True, field_statistics=True, offsets=False)\n",
    "    vector = term_vector['term_vectors']['TI']['terms']\n",
    "    f_stats = term_vector['term_vectors']['TI']['field_statistics']\n",
    "\n",
    "    # The answer should sum over 'tokens', check if the tokens exists in the 'vector',\n",
    "    # and if so, add the appropriate value to 'similarity'.\n",
    "    # Tip: add print statements to your code to see what each variable contains.\n",
    "    \n",
    "    similarity = 0\n",
    "\n",
    "    # BEGIN ANSWER\n",
    "    # END ANSWER\n",
    "    return similarity\n",
    "\n",
    "bm25_similarity(\"structure refinement\", 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# eventual comments here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See below the 'reference score' computed by ElasticSearch:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "body = {\n",
    "  \"query\": {\n",
    "    \"match\" : { \"TI\" : \"structure refinement\" }\n",
    "  }\n",
    "}\n",
    "explain = es.explain(index=\"genomics-base\", id=\"3\", body=body)\n",
    "print (explain['explanation']['value'])  # BM25 score computed by ElasticSearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "97cc609b13305c559618ec78a438abc56230b9381f827f22d070313b9a1f3777"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
